# Transcriptomics

[0. UvA SBiP](Biology/UvA%20Systems%20Biology%20in%20Practice/0.%20UvA%20SBiP.md)

## Transcriptomics


## How to determine expression

### Northern blotting

- Northern blotting
	- isolate RNA
	- run on a gel
	- radioactive label bind to RNA and visualized under x-ray

### qPCR

- qPCR
- Measure multiple genes at the same time
	- micro-array

### Micro-array

- steps:
- need to design probe for each gene intended to measure
- assumption: more RNA -> more binding
	- affinity between RNA and probe affects signal strength
		- low amount, high affinity -> high signal
- quantification is not absolute
	- upregulated or downregulated but not exact number of copies

### Sanger sequencing


### NGS

#### NGS: trick 1

- Spatial separation
	- Separate each unique piece of RNA
	- Fix in place
	- Amplify to improve signal
	- Sequence
- Adapter
	- Ligate the DNA
	- PCR amplification
	- Spread and fixate
	- -
	- -
- Emulsion PCR (Ion torrent):
	- beads and DNA
	- beads bind to DNA and placed into droplets
- Flow cell (Illumina)

#### NGS: Trick 2

- Real time sequencing
- with Ion torrent
- with Illumina

#### NGS Techniques

- Sanger:
	- Most accurate, slow
- Ion torrent:

#### NGS Workflow

isolate RNA, make cDNA


### RNA degradation

plots:
large peaks: ribosome subunits
in time, large peaks degrade into smaller ones
mRNA have different lengths and are spread everywhere

### Do we want to sequence all RNA

- eukaryotes: 

### mRNA enrichment

- risk: lose target while doing enrichment
- correction factor: lactate in water, enrich, find how much is retained
- instead, capture the target instead of getting rid of off-target
- another: poly-A tail; limit in eukaryotes that have a poly-A tail

### rRNA


### rRNA depletion

- available in kits

### RNA seuqncing

- Isolate rna
- Qualty check
- purifyh RNA
- Fragment
- convert to dsDNA
- Ligate adapters
- PCR 
- Quality check
- seuqncing in illumina

## Data analysis

### Workflow

1. Check quality of fragment sequences
2. Remove unwanted/low quality sequences
3. Map fragments to genome
4. Count number of fragment per feature (genes)

### Errors in sequencing

- Contamination (with other samples)

### PCR bias

- Some sequences are easier to PCR, some are harder (can be 1000x harder) ex. GC content high
- get rid of PCR bias: 
	- break RNA into pieces
	- hundereds of reads per gene
	- some regions have high reads. some have lower
	- compare the poorly sequenced region across conditions, not with other regions in the genome

### FASTQ file

- Sequencing output
- Line 1: 
- Phred: base call quality scores
- Cleaning:
	- Adapters
	- Low quality segments
	- PCR duplicates
	- Tools:
		- Trimmomatic
		- Samtools
- Mapping: 
	- From which genes did the fragments originate
	- fragments are small and referencees are large
	- same gene in multiple location in the genome -> multiple matches
	- actual genome is not exactly the same as the reference, affected by experimental condition

### Single and pair-ended


### Mapping; aligners


### Counting

$$
\begin{align}
R^2&=1-\frac{SS_{res}}{SS_{tot}}=1-\frac{\sum_{i}(y_{i}-\hat{y}_{i})^2}{\sum_{i}(y_{i}-\bar{y})^{2}} \\
\chi^{2}&=\sum_{i}^n \frac{(y_{i}-\hat{y}_{i})^{2}}{\hat{y}_{i}}
\end{align}
$$