# Metabolomics and proteomics

[0. UvA SBiP](Biology/UvA%20Systems%20Biology%20in%20Practice/0.%20UvA%20SBiP.md)

## Metabolomics Background

### How large is the metabolome

- Human metabolome
- Different metabolomes

### Small molecules are important

- >95% of all diagnostic clinical assays test for small molecules
- 89% of all known drugs are small molecules
- 50% of all drugs are derived from pre-existing metabolites
- 30% of identified genetic disorders involve disease of small molecule metabolism
- Small molecules serve as cofactors and signaling molecules to thousands of proteins

### Metabolism in body

- Inborn disorders of metabolism
	- Checked in babies by taking a blood sample from the heel
	- Early detection can be treated
- introduced different diseases

### Metabolites can show dramatic changes

- Metabolite changes fast in response to disturbance
- Quench in liquid nitrogen or use organic solvent, both stop metabolism
- Measured separately: polar and non-polar

### Metabolomics is the most complex omics

- A small change in the genome and cause large changes in metabolome via the proteome
- Genome: four letters
- Proteome: 20 amino acids
- Metabolome: over 3000 chemical classes
	- Measured with NMR, GC-MS, LC-MS

## Metabolomics workflow

### Steps

- Taking samples
- Extraction
- Biofluid of extracts, add buffer
- Chemical analysis
- Data analysis
	- With the question to answer in mind

### Many different metabolomics techniques

- Covers different part of metabolome
- More comprehensive studies use all of them

### Comparing omics coverage

- Genomics: 22,000 genes
- Proteomics: 5000 proteins
- Metabolomics: 200 chemicals can typically be detected, can be up to 800-1000's

## Proteomics background

### What are proteins

- Function comes from structure

### What is a proteome

- The total number of proteo-forms found in a cell or tissue
	- A protein can have different functional forms

### Why is proteomics important

- Proteins carry out cell metabolism
	- as enzymes that catalyze reactions
- Some diseases are disorders of protein function and causes downstream metabolomics changes
	- Multi-omics: regulation between different omics

### Typical proteomics workflow

- Similar to metabolomics workflow
- Many relevant analysis techniques
	- MS
	- 2D DIGE
	- Antibody based: 
		- detect specific proteins (ex. COVID detection kit)
		- low throughput
	- Interaction screens: yeast two-hybrid, other high throughput screening
	- NMR, X-ray diffraction, CryoEM: [2. VU SB Structure Determination](Biology/VU%20Structural%20Bioinformatics/2.%20VU%20SB%20Structure%20Determination.md), [6. VU SB Advanced Structure Prediction](Biology/VU%20Structural%20Bioinformatics/6.%20VU%20SB%20Advanced%20Structure%20Prediction.md)

### Proteomics is the most complex omics

- Different techniques generate different types of data
- Complex
	- > 500,000 protein forms
	- ex. post translational modifications of histones: proteins that binds the DNA, long chains and many modifications
- Dynamic range
	- The difference between the most abundant and the least abundant that can be detected above the noise level
	- ex. human plasma: 12 orders of difference in concentration from the most abundant albumin to less abundant immune markers; practically 5-6 orders can be measured

## M and P: differences and similarities

- Proteomics: more defined, structural is less known, but can be predicted now; not all modifications known; linked to genomics and transcriptomics
- Metabolomics: poorly defined; 

## General workflow in data analysis

### Sampling


### Pre-processing



### Further

- Normalize, impute, transform
- T test
- vendor and free software available: mzmine, xcms, maxquant

### Functional annotation

- GO, KEGG
- network analysis:
	- sometimes loose nodes (few edges) are removed, but they can be important information
	- [Metabolite annotation from knowns to unknowns through knowledge-guided multi-layer metabolic networking](Biology/Seminars/Metabolite%20annotation%20from%20knowns%20to%20unknowns%20through%20knowledge-guided%20multi-layer%20metabolic%20networking.md), [metDNA2 notes](LCMS/metDNA2%20notes.md)
- Most output is a matrix, sometimes a list
