# Mass spectrometry

[0. UvA SBiP](Biology/UvA%20Systems%20Biology%20in%20Practice/0.%20UvA%20SBiP.md)

## History

- Early MS design principle: bend ions; how much they bend depend on mass and charge (m/z value)
- DNA sequencing: enabled genomics and proteomics
- MALDI: proteomics
- Q-TOF: quadruple time of flight; proteomics
- Orbitrap: make ions circulate, circulating ions generate charge, charge is detected, frequency can be calculated
	- Fourier transform is used to deconvolute many ions circulating at the same time in frequency domain. Many ions can be measured at the sam time

## Proteomics 1

- Approach: bottom up and top down; mostly bottom up is used
- In bottom up approach proteomics sample prep mostly involves reduction and alkylation of cysteines followed by tryptic digestion
- The challenges of proteomics drove mass spectrometry development
- Proteome database search

### Where does it fit in systems biology

- Between genomics and metabolomics

### Top down bottom up and other proteomics

- b: peptide to protein
	- lose modification information such as phosphorylation
- t: from intact proteins, more for structural biology
	- contains modification information such as phosphorylation
- middle: make and analyze simpler proteins

### Top down

- Ion spray intact proteins 
- Complex protein:
	- fragment into subunits
	- measure subunits

### Bottom up

- Extract proteins
- Trypsin digestion
	- peptide too long can fall outside of detectable range
	- peptide too short is hard to map back to the protein sequence
- Purification: C18 column to remove buffer from digested protein solution
- LC-MS: more expensive C18 column to separate peptide fragments
- Spectra to annotation: in-silico fragmenter, database search

### 6 steps how to go from ions to protein identities

- phospho proteomics: need specific column to separate phosphorylated proteins

### Bottom up sample preparation

- Different proteins digest differently and extract differently. Relative comparison (A between sample x and y) can be made but not absolute (A and B within the same sample x)

#### steps


#### What we often do

- magnetic beads for protein to precipitate on
- can't use SDS, too strong they will contaminate the MS system

## Proteomics II

### Not all peptids are informative for identification and quantitation

- 

### Why is bottom up proteomics data incomplete

#### Limit of the duty cycle

#### Classical data dependent MS/MS

- MS2 is serial process
- MS2 is a biased process: only take the most intense peak first
	- why top down
- MS2 is a discontinuous process

### protein mixtures are incredibly complex

- longer gradient leads to more fragment spectra

### Dynamic range

- Size of detectable protein

## Protein level fractionation using SDS-PAGE

- Gel based analysis method
- Cut protein into multiple pieces -> become multiple samples
- Injection 

## Peptide levle fractionanation using HpH-RPLC

- 10x longer time

## Ion mobility separation: more separation dimension

- amoint of peptides
- retention time
- m/z value
- additional diemsnion: ion mobility, separate proteins based on movment speed in gas chamber, compact and small moves faster, heavy and large moves slowly
- TIMS: 

## Data independent analysis: DIA, MEe, SWATH

- Detect "everything"
- Loss of connection between precirsor and prodcut ions
	- do not know which precursor leads to the fragment
- For short gradients, DIA performs better than DDA (data dependent analysis)

## Prtoein identification depends on peptides

- mass window: 350 < x < 1750
	- too long can make movement when collide -> low fragmenetation issue
- protein sequences coverage: 56%
- coverage is not complete
- mass isotopologues:
	- neutral mass: no isotope
- need neutral mass to identify peptides
- steps:
	- in silico fragmentation; similar [metFrag](LCMS/metFrag.md)

## Prtoein infernece affects identification and quantitation



## how to go from ions to protein identities

seminar


## Quantitative prpoteomics

- Mass spectrometry measures the mass of ions

- label free, comapre in the end
- stable isotope label

## Applications of proteomics

- Expression proteomics: compare protein expression levels
- Regulation of phospho proteins: which pathways are activated
- Interaction proteomics: antibody, protein-protein interaction
- Protein analysis: confirm sequence, post-translational modification

## Expression proteomics can be quantitative

- Proteins last longer than mRNA


## Compartmental proteomics

- Break cell and do density gradients
- Quantitaively map marker proteins to cell location

## Interaction proteomis

- mRNA-protein interac