# Genome Sequencing

[0. VU Algorithms in Sequence Analysis](0.%20VU%20Algorithms%20in%20Sequence%20Analysis.md)
[5. VU FoB - Sequencing](5.%20VU%20FoB%20-%20Sequencing.md)

## The human genome project

- Dr. Craig Vector, Celera genomics
	- **shotgun method**: sequence in small segments then align them
- Francis Collins (USA), John Sulston (UK), Human Genome Project
- Jim Kent: coded the **GigAssembler**; saved the human genome project

### GigAssembler

Connect fragments based on overlaps
Use graph to connect fragments
When adding a new fragment, besides overlaps, also check the consistency with other fragments already in the sequence.

### HGP sequencing methods


### HGP numbers

- at that time it was difficult to sequence the DNA near the centromeres
- more than half are repetitive elements; this cause problems for reconstruction algorithms

## Next Generation Sequencing

[Illumina sequencing (NGS)](5.%20VU%20FoB%20-%20Sequencing.md#Illumina%20sequencing%20(NGS))

- shorter fragments than [Sanger sequencing](5.%20VU%20FoB%20-%20Sequencing.md#Sanger%20sequencing)
- very fast
- reduced cost
- massive parallel sequencing of fragments
- huge amount of data

## Sequencing (or assembly)

**The problem**: reconstructing a DNA sequence from many randomly selected short fragments; read may contain errors.

Two main ways of assembling the genome:
- De novo assembly of a genome
- Assembly using alignment onto a **reference genome**

### Assembly

- Read
- Paired end reads
- Raft
- Contigs

### Contig

shortest superstring problem
Contig: a continuous set of overlapping sequences

### Two problems

- multiple contigs, due to lack of overlapping reads
- repeats

### Depth of coverage measure


### Paired-end reads

- deal with gaps in contigs
- mate pair

#### Combining contigs with paired-end reads


#### identify repeat reads


## Carrying out de novo assembly

### Finding the Eulerian path

Bridges of Konisberg


### Directed graph

[Directed Graph](Directed%20Graph.md)


### K mers

### De Bruijn graph

- Nodes are k-mers
- Edges are (k+1)-mers, connecting 2 nodes
- Nodes need to have an overlap of (k-1) symbols
- Directed edges, such that the suffix of the outgoing node overlaps with the prefix of the incoming node

## Sequencing if a reference genome is available

Input: reference genome and a set of reads (**BAM** file)
Burrows Wheeler transform

Problem of mismatches: increase processing time compared with 

Burrows-Wheeler Aligner (BWA)

