# Microbiome Data Analysis

[0. VU BDA](0.%20VU%20BDA.md)

## Overview

- What are microbiomes and how do we analyze them?
- Challenges in data analysis
	- what causes them
	- how to recognize common problems
	- data transformation and normalization
- Microbial ecology data analysis methods for distance between samples

## Microbiome

- Microbes are essential partners of many animals and plants
- Many types of microbes
- Microbes often live in a community with other microbes
	- Many are hard to isolate and culture
	- E. Coli. can be cultured independently
- Study microbiomes
	- Studied by in-situ multiplex methods, such as meta-omics, aka. meta-barcoding
	- Steps
		- Sampling + preservation, such as stool sample
		- Sample processing
		- Sequence the genome: contain genome of host and microbes
		- Match sequence to taxonomic annotation
		- Count sequences to quantify bacteria taxa
- Microbiome research questions
	- What microbes are present
		- Pathogenic microbes
		- Anaerobic or aerobic microbes
		- All branches of the phylogenetic tree or a limited range
	- How similar are two microbiomes?
		- What microbes are shared?
		- Relation to phenotype
		- Cross-section: between different people
		- Over-time: compare microbiomes at different time points
	- Examples
		- Measured microbiomes for various body parts
			- which microbiomes are composed of similar bacteria
			- clustering
		- Measured microbiome of babies born vaginally or via c-section
			- classify which way the babies were born 
			- random forest
		- Measured microbiome of plants under drought and control conditions
			- what bacteria are more abundant in stressed plants
			- LDA or PCDA
			- multiple t tests
			- Features with high loadings in the direction that separates the 2 conditions
			- Feature importance analysis for random forest

## Challenges

### Identification of microbes

- a bioinformatics problem to identify microbes in a sample
- Methods with low or high false positive rates lead to very different results

### Sparsity

- Most microbes are present in a few sample
- A few microbes are present in all samples
- The matrix is [sparse](Sparsity.md), many 0s
- Reason
	- No underlying constraint unlike transcriptome or proteome (limited by genome)
	- Very different ecosystems: healthy has anaerobic microenvironment and inflamed gut has aerobic environment

### $\alpha$-diversity

- Aka. richness
- The number of taxa in a sample
- Can have ecological meaning

### Limited sampling depth

- The human gut microbiome, for example, is made of many bacteria taxa. Some of them make up a high percentage, while others make up a small amount
- **Question**: is the sequence depth deep enough to see all bacteria taxa?
- Rarefaction curve:
	- species vs. reads plot
		- how many reads does it take to see $y$ number of species?
		- Higher reads lead to higher diversity
- **Rarefaction**: a **normalization** technique to keep the number of reads the same in every sample
	- sequencing depth varies between samples due to technical variation
	- determine a rarefaction threshold: choose a sequencing depth to which all samples will be rarefied, usually based on the sample with the lowest read count
	- random sub-sampling: choose reads randomly for each sample up to the threshold
	- makes **diversity** comparable
	- suitable when:
		- comparing $\alpha$-diversity
		- samples have large differences in sequencing depth
		- no other normalization method is available
	- Pros
		- Removes bias caused by different sequencing depths
		- Standardizes data for fair comparison of microbial diversity
		- Prevent artifacts from over-sampled sequences affecting diversity estimates
	- Cons
		- Loss of data: reads are discarded
		- Introduces randomness: random in which reads are kept
		- Not suitable for abundance comparisons: alters relative abundances, which may distory biological signals
	- alternatives without discarding data: 
		- DESeq 2 normalization (variance stabilizing transformation)
		- Relative log expression (RLE)
		- Cumulative sum scaling (CSS)
		- Proportional normalization (converting counts to relative abundances)
	- why does gene expression data not need rarefaction
		- lowly expressed genes can still be detected

### Compositionality

- The microbiome data is compositional
	- Total number of reads cannot be used to quantify microbial abundance due to varying sequencing depth. Therefore, relative abundances of bacteria are analyzed.
	- They must sum to 100%
	- If another species is added, all the proportions will shift
- Compositional data consist of vectors whose components are proportions: $$x = (x_1,x_2,\ldots,x_s,\ldots,x_D)$$, and the proportions are constrained to a constant (unit-sum-constraint): $$x_1+x_2+\ldots+x_s+\ldots+x_D=1$$
- Problems:
	- Variables will be correlated
		- When one variable increases, another must decrease
		- Tests that assume **independence of variables** will produce misleading result, such as Pearson correlation
	- Euclidean distance is inappropriate
		- Measures distance between two samples
		- **Spurious correlations**: increase in one taxon does not necessarily mean a real biological increaseâ€”it might just be a side effect of other taxa decreasing.
		- The same absolute change in two taxa might _mean different things_ depending on their original proportions.
			- A change from 1% to 5% is much more significant than a change from 40% to 44%, but Euclidean distance treats both as the same (both are a +4% increase).
	- Log transformation can fail
		- Many models assume **normally distributed** data, but microbiome data is **right-skewed and zero-inflated**.
		- Direct log-transformation is problematic due to zeros in the data.
- **Centered log-ratio (CLR) transformation**:
	- transformed **compositional** data, such as microbiome relative abundances, into an **unconstrained** space where standard statistical techniques can be applied.
	- For a composition $x = (x_1,x_2,\ldots,x_D)$, the CLR transformation is: $$CLR(x_i)=\log(\frac{x_i}{g(x)})$$, where $g(x)$ is the **geometric mean**, defined as: $$g(x)=\left(\prod_{i=1}^Dx_i\right)^{\frac{1}{D}}$$
	- After CLR transformation, analyses using Euclidean distance such as PCA and clustering can be applied
- Transformation does not recover lost information from converting to relative abundances
	- say relative abundance of A is 50% in sample 1 and 20% in sample 2, even after transformation, we still do not know whether it's because whether there are more A in sample 1 or just less other species in sample 1.
	- Need to measure absolute quantities, such as ng DNA/mg of stool sample to determine.
- Problem with CLR: many **zeros** in the data
	- not good for calculating ratios and taking logs

### Ecological perspective on microbiome study

- Many other methods discussed in this course are not used in ecology for 2 reasons
	- Ecologists ask different questions
	- Different interpretation of 0s

#### Ecology questions

- Traditionally, ecology focuses on the relationship between samples and not on which species differ between samples.
	- there are too many bacteria that are different
- Or how samples relate to environmental factors

#### Interpretation of 0s

- Why do we observe 0s: not present or not detected
- Interpretation of double 0s
	- a species is not observed in 2 samples
	- **environment gradient**: a spectrum of pH, temperature, food availability, etc.
	- a species is not present can be because the environment gradient is too high or too low
		- other reasons: dispersal limits, priority effects, recent extinction by virus/predator
	- or that a different species has occupied the niche
	- or that the environment is suitable, but the species has not arrived
	- or that it is present but not detected
	- key point: double 0s contain **no ecological information**
		- or the absence of 1 species from 2 samples does not make the two samples more similar
- How _dissimilar_ are 2 samples that have no common taxa?

## Microbiome analysis methods

- Ask microbiome questions
- Have microbiome interpretation of 0s

### $\beta$-diversity

- Distance/dissimilarity between samples
- Question: what is considered similar
	- Depends on your question
	- To illustrate, compare sample A, B, and C below. Which samples are more similar?
		- A and B have no bacteria 1, but A and C have close relative abundances of bacteria 2 and 3
		- Ecologically, C is maximally different because there is no common species with A and B
		- However, in PCA, C and A are positioned together. They have higher euclidean distance to B.
		- Hence, euclidean distance is not a good choice

| Taxon | Bact. 1 | Bact. 2 | Bact. 3 |
| ----- | ------- | ------- | ------- |
| A     | 0       | 1       | 1       |
| B     | 0       | 4       | 8       |
| C     | 1       | 0       | 0       |
- **Aitchison distance**: 
	- One of $\beta$-diversity measures
	- euclidean distance of the CLR data: $$d_A(\mathbf{x}_i,\mathbf{x}_j)=\sqrt{\sum_{s=1}^D\left[\ln\frac{x_{si}}{g(\mathbf{x}_i)}-\ln\frac{x_{sj}}{g(\mathbf{x}_j)}\right]^2}$$
	- This way, similar samples are closer together
	- Need to add pseudo-counts for log transformation: add 1 to 0s
	- This means that all 0s are interpreted the same way. Hence, samples that have 0s in similar positions will be positioned **together**, even if 0s mean they _do not share the same species_. 
- **Bray-Curtis distance**:
	- Another $\beta$-diversity measure
	- Sum of absolute value of difference divided by sum of total: $$d_{BC}(\mathbf{x}_i,\mathbf{x}_j)=\frac{\sum_{s=1}^D|x_{si}-x_{sj}|}{\sum_{s=1}^D(x_{si}+x_{sj})}$$
	- If the numerator and the denominator are the same, $d_{BC}=1$, and the two samples are completely dissimilar
	- This way, samples that are dissimilar are maximally far from each other
- **UniFrac distance**:
	- If two organisms that are closely related to each other, they are also more similar.
	- Imagine a case: sample A contains 3 species, and sample B contains the same 3 species and a 4th species closely related to the other 3. We can still say A and B are similar.
	- In short, two samples can be considered similar if they have **related species**
	- Species relatedness is measured using a phylogenetic tree, which can be constructed by analyzing the sequenced DNA. The more different the DNAs are, the further they are in the tree of life.
	- **Unweighted**: $$u=\frac{\sum_{n=1}^Nl_n\left|A_n-B_n\right|}{\sum_{n=1}^Nl_n\max\left(A_n,B_n\right)}$$
		- $N$ = Number of branches in the phylogenetic tree.
		- $l_n$ = Length of the n-th branch in the phylogenetic tree.
		- $A_nâ€‹, B_n$â€‹ = Presence (1) or absence (0) of the $n$-th branch in samples $A$ and $B$.
		- Numerator: unique phylogenetic distance
			- total branch length that is unique to either A or B
		- Denominator: total branch length observed in at least one of the communities
		- ranges from 0 (same) to 1 (totally different)
	- **Weighted**: $$w = \frac{\sum_{n=1}^{N} l_n \left| \frac{A_n}{A_T} - \frac{B_n}{B_T} \right|}{\sum_{s=1}^{S} L_s}$$
		- $N$ = Number of branches in the phylogenetic tree.
		- $l_n$ = Length of the n-th branch in the phylogenetic tree.
		- $A_n, B_n$ = The abundance (relative count) of the microbial taxa associated with the n-th branch in communities **A** and **B**.
		- $A_T = \sum A_n$ = Total abundance (sum of all taxa abundances) in community **A**.
		- $B_T = \sum B_n$ = Total abundance (sum of all taxa abundances) in community **B**.
		- $S$ = The total number of branches considered.
		- $L_s$ = The root to tip length of a branche
		- Numerator: abundance weighted phylogenetic difference
		- Denominator: total sum of branch lengths in the phylogenetic tree
		- In short, weighted UniFrac distance is determined by _relative abundance_ and _phylogenetic distance_.
		- Also ranges between 0 (same) to 1 (different)
	- **Important consideration**: for weighted UniFrac distance, if the sequencing depth is very different between 2 samples, they will probably miss some rare species. If these species also have a far evolutionary distance, these two samples will have a large UniFrac distance, which may or may not reflect the biological truth. In this case, **rarefactioning** should be done to make the sequencing depth the same.

### **Comparison: Weighted vs. Unweighted UniFrac**

| Feature                          | **Unweighted UniFrac**                          | **Weighted UniFrac**                       |
| -------------------------------- | ----------------------------------------------- | ------------------------------------------ |
| **What it Measures**             | Shared phylogenetic lineages (presence/absence) | Relative abundance differences             |
| **Sensitivity to Rare Taxa**     | High                                            | Low                                        |
| **Sensitivity to Dominant Taxa** | Low                                             | High                                       |
| **Use Case**                     | Detecting community **membership differences**  | Detecting community **composition shifts** |
### Representing $\beta$-diversity matrix

- $\beta$-diversity matrix: a matrix containing $\beta$-diversity between each pair of samples
	- A distance matrix
- The samples can be plotted using principal coordinate analysis (**PCoA**) in an Euclidean space.
- Note that the distance matrix does not preserve how the original variables contribute to the distances, i.e. does not produce loadings like PCA
- Steps of PCoA:
	- transform matrix: $A = -\frac{1}{2}D^2$
	- center rows and columns (the matrix is symmetric so rows and column centering are the same)
	- calculate eigenvalues and eigenvectors
	- multiply eigenvectors by square root of eigenvalues
	- the product is the principal coordinates
	- **Note** this is similar to the eigendecomposition way of doing PCA
- Can be done for any distance matrices mentioned above: Aitchison, Bray-Curtis, UniFrac, and euclidean. 
	- For euclidean distance, the result is the same as PCA of original data. 
	- For Aitchison distance, the result is the same as doing PCA for CLR transformed data.

## Testing effects on $\beta$-diversity: PERMANOVA

- Statistically test if two groups are different or not
- Divide data into groups based on factors, such as treatment condition
- Calculates between group and within group variation using distance measures
	- Total: $$SS_T=\frac{1}{N}\sum_{i=1}^{N-1}\sum_{j=i+1}^{N}d_{ij}^2$$
	- Within: $$SS_W=\frac{1}{n}\sum_{i=1}^{N-1}\sum_{j=i+1}^{N}d_{ij}^{2}\in_{ij}$$
	- Between: $$SS_B = SS_T-SS_W$$
	- F-statistic: $$F=\frac{\frac{SS_B}{groups-1}}{\frac{SS_W}{N-groups}}$$
	- Permute (swapping labels) and calculate multiple F statistics to form a distribution.
	- Calculate the p value of the observed F statistic using the distribution.