# Mutation, Evolution, and Sequence Alignment

Same topic as [2. MIT CompBio - Dynamic Programming](2.%20MIT%20CompBio%20-%20Dynamic%20Programming.md) and [3. MIT CompBio - Hashing, Database Search, BLAST algorithm](3.%20MIT%20CompBio%20-%20Hashing,%20Database%20Search,%20BLAST%20algorithm.md)

## Mutation

- Missense mutation: a mutation that changes an amino acid in the protein
- Nonsense mutation: a mutation that changes an amino acid to a stop codon

- Insertion/deletions: cause frame shifts
	- Inserting multiples of 3 does not cause frame shifts
	- Each codon is 3 nucleotides long
- Transitions and transversions
	- Transition: between purines or pyrimidines
	- Transversions: across purines and pyrimidines
- [Formulation 3: Sequence alignment](2.%20MIT%20CompBio%20-%20Dynamic%20Programming.md#Formulation%203%20Sequence%20alignment)

## Evolution

- Convergent evolution: different species arrive at the same features in evolution, but evolved in different paths.
- Adaptive radiation/divergent evolution: homologous structure evolving into different forms.

### **Homology**

- sharing a common ancestor
- two homologous sequences are evolutionary related
- normally share function and structure

## Sequence alignment

### Substitution matrix for amino acids

- Defines substitution cost scores
- PAM (Point of Accepted Mutation)
- BLOSUM (Block Substitution Matrix)
	- BLOSUM 62
	- Contain log odds ratios of amino acid substitutions

### Odds ratio and log odds

$$
\begin{aligned}
P_{success} &= P \\
P_{failure} &= 1-P \\
Odds &= \frac{P_{success}}{P_{failure}} \\
LogOdds &= \log(\frac{P}{1-P})\\
\end{aligned}
$$

Log odds
- 0 means 50% chance will happen
- Positive value means likely to appear
- Negative value means less likely to appear

Read the rest on [Dynamic Programming for Sequence Alignment](2.%20MIT%20CompBio%20-%20Dynamic%20Programming.md#Dynamic%20Programming%20for%20Sequence%20Alignment)

## BLAST

- Basic Local Alignment Search Tool
- Alignment algorithm

### How it works

- Input: query sequence
- Database: contain many sequences to be matched
- Pigeonhole principal
- Break query into k-mers
- Extend to neighborhood of similarity
- Search through DB and store matrix
- Seed and extend local alignment

PSI-BLAST