# Integrative analysis of single cell and spatial omics

#paper 
[0. VU B4TM](Machine%20Learning/VU%20Bioinformatics%20for%20Translational%20Medicine/0.%20VU%20B4TM.md)

> Ahmed Mahfouz, AP, Leiden MC
> Pattern recognition and bioinformatics, TU Delft
> mahfouzlab.org

## Spatially cmapped transcriptomics

- In situ hypridization (ISH)
- 3D gene expression: 20k genes, 60kvoxels
- 3700 samples: microarray, RNA-seq

## Understrnading ...

- spatial temporal patterns
- regulatory networks
- embedding/visualization
- integration with imaging data

## Increased resolution using sc/snRNA-seq


## Spatial transcriptomics

- NGS based: dots with barcodes to label RNA. The barcodes label spatial location
	- problem: limited resolution
- In-situ hybridization: colored barcoding, many rounds
- In-situ sequencing: rolling circuit ampilification

## Background

- DMD gene
- 79 exons, 7 promoters, 8 major isoforms
- Duchenne Muscular Dystrophy: caused by mutation, leads to shorter life expectancy
- Open questions:
	- mechanisms driving muscle pathology
	- DMD function in brain
	- -
	- -

## Applications of spatial sequencing

### Discovery: Histopathology markers in DMD

- Spatial labeling of non-fibrotic and fibrotic sections
- Identify markers in each region

### Discovery: 


### Discovery: Effects of lacking MD isoforms in the brain


### Targeted: effects of treatment on mouse brain

- Wrong splicing causes reading frame shift that does not translate to protein
- Exon skipping: skip an exon that leads to a less function protein
- Gene replacement: replace with a less functional version of the protein
- Still do not know what the protein does in the brain
- Xenium panel: 
	- Probe hybridizes to transcript. Circular amplification happens when the transcript is there and the probe is bound to it. The amount of signal indicates the amount of transcripts

## Different technologies answer different questions

- sc RNA seq: full transcriptome, no spatial information
- NGS based: full transcriptome, low resolution
- in-situ sequencing/x-FISH: high resolution, limited genomic features

## Single cell + spatial

- Predict the location of dissociated cells and deconvolution
- Predict spatial expression of unmeasured genes

## SpaGE: spatial gene expression enhancement

- ref scRNA data
- queryL spatial data
- KNN imputation of gne expression in spatial data from scRNA data

### probelm : single cell and spatial data don;t align

- data from two experiments
- no alignment
- domain adaptation using PRECISE
	- align the PCs of the two datasets
	- minimize cosine difference

### SpaGE in primary visual cortex (VISp)

- Data
	- STARmap: 1549 cells, 1020 genes
	- scRNA seq: 14249 cells, 34617 transcripts
- SpaGE predicted expression in STARmap

### Indenpendent benchmarking of SpaGE

- Compared to other methods

## What else can we transfer from scRNA-seq to spatial data

### RNA velocity

- In scRNA sequencing, probe binds to different location than their intended binding site sometimes
- mRNA is spliced and degraded in time. This leads to a concentration change of mRNA, forming a curve
- RNA velocity maps the cell to a point on the curve. It shows in some time, the current cell will become like which cell
- Not applicable for spatial data:
	- -
	- -

### SIRV: Spatially inferred RNA velocity


### Evaluation of SIRV estimated velocities

- MERFISH data

### Applying SIRV to the developing mouse brain

- Inferred RNA velocity matches with known trajectory

## Information

- Sptatial
- ELIXIR-SCO-spatial-omics
- a.mahfouz@lumc.nl
- genomic language models
- cell perturbation prediction
- spatial data analysis