# RNA sequencing

[0. VU B4TM](Machine%20Learning/VU%20Bioinformatics%20for%20Translational%20Medicine/0.%20VU%20B4TM.md)

## RNA seq workflow and data pre-processing

### Mapping and quantification

- Count how many reads map to each gene
	- Output: count matrix (gene x sample)
	- higher count implies higher expression

### Common RNA-seq analyses

- Differential gene expression
	- Which gene differ between conditions
	- edgeR, voom-limma, DESeq2
- Clustering/subgroup identification
	- Mamaprint, PAM50

### Differential gene expression

- Which genes have a different expression between groups
	- Drug vs. control
	- Mutated vs. WT
	- Disease vs. healthy
- Is there a different in mean between two groups?
	- Count data not normally distributed: need normalization
	- Need normalization (eg. scale by library size)
- More here, add later

## Considerations for differential gene expression

### Biases in RNA-seq: 

> what effects determine how many reads I count for gene A in sample X

- Expression of gene A in sample x (what we want to measure)
- Library size (more reads from sample -> more reads per gene)
	- Can be normalized using counts per million, but in practice typically more sophisticated normalization is used.
	- Another approach: assume most genes are not differentially expressed, look for a factor that multiplies the library size and  minimizes the difference between the samples
- Transcript length (longer genes -> more reads)
	- this is also known and can be accounted for
- Mappability of reads
- Position: fragments are preferentially located towards beginning or end of transcripts

> Implication: bias of a gene is the same across samples, but biases of genes are not the same within the same sample. This means that gene expression is a relative measurement, and that we can compare gene A between samples X and Y but **not** gene A and B within sample X.

## Example: walk through differential gene expression

### Biological background

- RNA-seq of $SHP_{2}$ xenografts
- $SHP_{2}$ is important for signal transduction and plays a role in KRAS mutant cancer
- Mice were implanted with cancer cels with/without $SHP_{2}$ (=PTPN11)
	- $PTPN_{11}$ WT and $PTPN_{11}$ KO
		- KO with CRISPR
	- Two independent KO clones (B17 & B33)
		- To confirm difference observed in one clone is caused by the KO in the other clone
	- 3 replicates per condition
- 3 groups in total: WT, B17 KO, B33 KO
- RNA-seq on resulting tumors

### Dimension reduction

- For a look at the data
- Methods: MDS, PCA, UMAP
- Idea: project the many genes to a lower dimensional space
	- more similar points are closer to each other
- What would it mean if the WT and KO samples are all mixed?
	- No difference between samples
	- Very few genes are significantly different, could be due to chance

### Effect size

$$
\log FC = \log_{2} \frac{\text{expression condition A}}{\text{expression condition B}}
$$

$$
\log_{2}\left( \frac{KO}{WT} \right)=-\log_{2}\left( \frac{WT}{KO} \right)
$$

- log fold change makes differences better comparable than absolute differences
- Also allows fair comparisons of genes with very low expressions
- Normal distribution

### Differential expression is measured by statistical significance and effect size

- [Definition of the p value](Machine%20Learning/VU%20Biosystems%20Data%20Analysis/2.%20VU%20BDA%20Preprocessing.md#Definition%20of%20the%20p%20value)
- Smaller p value = more significant
	- often 0.05 is used as cutoff
- Depends on:
	- difference between the groups
	- difference within groups
	- the number of samples
- More samples gives higher significance (lower p values)
	- but not necessarily biologically significant
	- check effect size

### When is a gene differentially expressed?

1. p: $3.6*10^{-5}$, logFC: $-6.2$
2. p: $0.003$, logFC: $-0.38$
3. 0
4. 0

### The mean-variance relationship: borrowing information between genes gives a better variance estimate

- how accurate is the variance estimate?
- for smaller sample size, the place of one point can affect greatly the variance estimate
- higher reads leads to smaller standard deviation -> better variance estimate
- highly expressed genes tended to have lower p values -> more confidence in its expression value
	- small expressed genes can be affected more profoundly by noise

### Multiple hypothesis testing

- How many genes will be "differentially expressed" **by chance** in 20,000 genes if we use a 0.05 cutoff?
	- $20,000 \times 0.05=1000$
- Solutions: controling for multiple testing
	- $\alpha$: cutoff value used for significance (typically 0.05)
	- False discovery rate (Benjamini-Hochberg): at most $\alpha \%$ of yout hits are false positives
		- `p.adjust(method='bh' or 'fdr')`
	- Bonferroni: $\alpha \%$ chance that any of the hits is a false positive (much stricter!). 
		- Hit: p value less than $\frac{\alpha}{\text{number of genes}}$ 
		- `p.adjust(method='bonferroni')`
	- which one to use depends on what the downstream task is. Typically BH is used. Bonferroni could be too strict when the feature number is very high.


