# Functional enrichment of alternative splicing events with NEASE

[0. VU B4TM](Machine%20Learning/VU%20Bioinformatics%20for%20Translational%20Medicine/0.%20VU%20B4TM.md)
#paper 

## Alternative splicing

- Exons
	- Untranslated region
	- coding sequence
- introns
- Alternative splicing: different combination of exons lead to different proteins
- Implications
	- Protein functional difference
	- Affects PPI
	- Rewire PPI networks
- Detect alternative splicing: RNA sequencing

## Generations of sequencing

- Sanger sequencing
- NGS: Illumina
- Single cell sequencing

### RNA seq for alternative splicing

- Calculate how many reads correspond to exon $c$ 
- percent spliced: $\frac{a+b}{2c+a+b}$

### Long reads vs. short reads

- Short reads: hard to map to original genome, gaps may be present
- Long read: more expensive, low coverage

## Differential alternative splicing analysis

- Optimal tool does not exist yet
- Low recall: cannot identify all splicing that is present

## Enrichment

- Functional description of genes: reactome, KEGG, Gene ontology
- Steps:
	- Take significant genes (>0.05 and up/down regulated by a certain fold)
	- Background genes
	- Map to gene set: in gene set, not in gene set
	- Statistical test: FIsher's exact test, chi-square test

### Beyond Fisher's exact test (a side note)

- Choosing a threshold can introduce 
- Threshold dependent: Fisher's exact test
- Whole distribution: GSEA

## Main idea

- Alternative splicing event (ASE) leads to
	- Co-expression
	- ASE-driven network remodeling
	- Protein abundances, protein-protein interactions
- Understand disease mechanisms

## Problem 1: PPI neglects splicing

- One isoform of a protein interacts with another protein, but its isoform may not
- One isoform of a protein does not interact with another protein, but its isoform may do
- DIGGER database:
	- combines domain-domain interaction and protein-protein interaction
	- if the interacting domain is lost, the interaction becomes weaker or lost

## Problem 2: enrichment analysis operates on a gene level

- Splicing is neglected. A gene can lead to different proteins due to alternative splicing
- NEASE

## All problems solved?

- Domain loss may not lead to interaction loss
- DIGGER and NEASE cover a small portion of all interactions
- New organisms
- Using long read data: more accurate reconstruction of transcripts

## DIGGER 2.0

- More interactions based on homology (in other organisms)
- Interactions based on the predictions (PPIDM method)
- New model organism: Mus Musculus

## Projects

- isoform function inference for splicing-aware enrichment
- splicing aware knowledge graph
- DL/ML for isoform-isoform interaction prediction



