# Cancer phosphoproteomics

> Connie R. Jimenez
> Amsterdam UMC

#paper 
[0. VU B4TM](Machine%20Learning/VU%20Bioinformatics%20for%20Translational%20Medicine/0.%20VU%20B4TM.md)

## Background

- Cancer: complex and multi-factorial
	- environment, diet, lifestyle, genetic vulnerabilities
- Proteome: integrative read-out of the factors
- Mass spectrometry: proteome and interactome
	- 100+ MS/MS per second
	- 8000+ proteins per hr

## Proteomics: multi-disciplinary effort

- Sample preparation
	- sections, biopsy, biofluid
	- cell lysis, protein digestion
	- peptides
- Separation
- Data acquisition
	- DDA/DIA
	- MaxQuant/DIA-NN to extract peptides
- Data matrix
- Data science

### Proteomics applications

- biomarker discovery by --
- protein signaling by 
- protein function by interaction proteomics
- antigen discovery by _immunopeplidomics_

### International cancer proteomics CPTAC

- CPTAC: clinical proteomic tumor analysis consortium
- Integrative proteogenomics addis missing biology
	- Tumor specific MS/MS database
	- Prioritize 

## CNA regions and candidate drivers

- Correlation between CNA (copy number) regions and mRNA expression
- Helps narrow down --

## Pan cancer data


## Phospho proteomics

- Protein kinases important for protein function -> cancer
- Identify drugs: kinase inhibitors
- Drug resistance: need to identify escape pathways
- Pioneer study: global survey of phosphotyrosine
	- different cancer types reveals heterogenieity in phosphotyrtosine epxfession

### Affinity purification

- mass jump of 80 daltons between fragments

### Ranking approach

- ALK RTK
- Can be used to stratify patient