# VU ISB Practical 3

[0. VU Introduction to Systems Biology](0.%20VU%20Introduction%20to%20Systems%20Biology.md)

## E. coli genome sequence

```
GTGATAGTTGACGAGTGAAATTATGAACGTGCAAAACAAATCGCCGTTACCAGACAAATAATGCATTTACTGGAGAATTAATCCCGCTGTACGGAAGTGTGACGCAGGTCGACCTTGTGAGGGGACACAAAGAAAAATTGAATAAACTGTATGATTTAAAAGATTATCGGGAGAGTTACCTCCCGATATAAAAGGAAGGATTTACAGAATGTGACCTAAGGTCTGGCGTAAATGTGCACCGGAACCGAGAAGGCCCGGATTGTCATGGACGATGAGATACACCGGAATATCATGGACATATTCTTTAAAGCGCCCTTTATCTTCAAATGCGGCACGGAAACCGGAGGCTTTGAAGAACTCAAGGAAGCGCGGCACGATACCGCCCGCAATAAACACGCCGCCAAATGTCCCGAGATTGAGCGCCAGATTGCCGCCAAAACGGCCCATAATGACGCAAAACAGCGACAATGCGCGGCGGCAATCGGTGCAGCTGTCAGCCAGCGCGCGTTCGGTAATATCTTTTGGCTTGAGATTTTCTGGCAGGCGGTTGTCAGCTTTCACAATTGCGCGATACAAATTCACCAGCCCAGGGCCAGAAAGCACGCGCTCCGCCGAAACATGACCAATTTCCGCACGCAATATTTCGAGGATAATGGCCTCTTCTTCACTATTCGGCGCAAAATCAACGTGACCGCCTTCGCCTGGCAAGCTTACCCAACGCTTATCGACATGGACCAGATGCGCAACCCCAAGCCCCGTTCCGGCACCGTAAACCGCAATAGGCTTACCTTCGACCGGTTCTGCGCCACCAAACTGAATCAGATGCTCTTTTTTCAGCATCGGGATCGCCATCGATACAGCGGTAAAATCGTTAATAATTTCCAGATGGCTAAAACCGAGATTCTTTTTCATTTCGGCAATTGAGAACGCCCAGGTATGGTTGGTCATCGCCACCCAGTCACCGGTAATTGGGCAAGCGATGGCAATACAGCCGTCTTTCACCTCGACCTTATGTTCTTCAAGATAAACGCGAATGACCGCTTCGAGGCTGGGGTAATCAAGCCCTGAATAGGTCTTAGCCTGCGAGATTTCACCACTGGCAATATCACACAGAGCAAGACGTGCGTTGGTGCCGCCCACATCACCGACTAATGCATACTTTGTCATTCTTCAACTGCTCCGCTAAAGTCAAAATAATTCTTTCTCACACTGTAAATACCTGGGGGCATAACAACAACGCCGGAAAGGCAGGCTCCCTGTAAATATCGATCTGGGTCACA
```

```
>lcl|ORF7
MTKYALVGDVGGTNARLALCDIASGEISQAKTYSGLDYPSLEAVIRVYLE
EHKVEVKDGCIAIACPITGDWVAMTNHTWAFSIAEMKKNLGFSHLEIIND
FTAVSMAIPMLKKEHLIQFGGAEPVEGKPIAVYGAGTGLGVAHLVHVDKR
WVSLPGEGGHVDFAPNSEEEAIILEILRAEIGHVSAERVLSGPGLVNLYR
AIVKADNRLPENLKPKDITERALADSCTDCRRALSLFCVIMGRFGGNLAL
NLGTFGGVFIAGGIVPRFLEFFKASGFRAAFEDKGRFKEYVHDIPVYLIV
HDNPGLLGSGAHLRQTLGHIL
```
1. How would you know which part of this sequence codes for a protein? What are all the factors to consider when searching for this information? Hint: there are multiple ways to read this sequence, consider the double strand.
	1. Both directions, 3 reading frames, total of 6
		1. When reading backwards (3' to 5'), the codons are read from the complementary strand to this strand.
	2. Look for the start and stop codon
2. [ORF finder](https://www.ncbi.nlm.nih.gov/orffinder): looks for open reading frames in the DNA sequence
	1. 7 found
	2. The longest has 966 nucleotides/321 amino acids in the reverse direction
3. The longest sequence can be sent to BLAST from the ORF finder result page.
4. The protein is glucokinase. It is in multiple bacteria species, including E. Coli.
5. Look up this protein in the EcoCyc database. What is its function? In the cellular overview, use the highlight function (menu on the right) to see where it is acting.
	1. It is involved in the glycolytic process (GO terms tab)
	2. The cytoplasmic glucokinase is not required for growth on glucose as the carbon source, because glucose is phosphorylated during transport by the PTS system. Glucokinase phosphorylates glucose which is produced by amylomaltase.
	3. Find `Show on Cellular Overview` in the right-hand menu
		1. Highlight by enzyme: glucokinase
		2. Highlight by gene: glk
		3. Same highlight in the carbohydrate degradation pathways
		4. Tick checkboxes on top right to add and remove highlights
6. In EcoCyc genome browser, look at the region of the genome that codes for this protein. Look at what is coded in the neighborhood of this protein, are there any patterns?
	1. In the gene search result page, click on `View in Genome Browser`
	2. In the gene search result page, click on `Operons` tab
7. On the overview page for the protein, follow the link the UniProt database (the link is in the session "Accession IDs" of the protein page), where you can see even more annotations on this protein, including structural information. How was the structure of the protein obtained?
	1. In the gene search result page, click on the UniProt link in `Accession IDs`
	2. In the Feature viewer, view protein structure
		1. Molecule processing/chain: Extent of a polypeptide chain in the mature protein
		2. More information by searching in the HELP tab
	3. ALPHAFOLD predicted structure
8. Back at EcoCyc, look at the reaction of which this protein is a part of. Clicking in one of its substrates or products you can see a list of all the reactions they are a part of. Are all these connections shown in the "Cellular Overview"? Highlight all the reactants in cellular overview in different colors. This shows you how connected metabolic networks are.
9. Consider ATP. Show it on the cellular overview map. Why is it such an important reactant. Which other reactant do you expect to be so important?

## Mutated glk

```
GTGATAGTTGACGAGTGAAATTATGAACGTGCAAAACAAATCGCCGTTACCAGACAAATAATGCATTTACTGGAGAATTAATCCCGCTGTACGGAAGTGTGACGCAGGTCGACCTTGTGAGGGGACACAAAGAAAAATTGAATAAACTGTATGATTTAAAAGATTATCGGGAGAGTTACCTCCCGATATAAAAGGAAGGATTTACAGAATGTGACCTAAGGTCTGGCGTAAATGTGCACCGGAACCGAGAAGGCCCGGATTGTCATGGACGATGAGATACACCGGAATATCATGGACATATTCTTTAAAGCGCCCTTTATCTTCAAATGCGGCACGGAAACCGGAGGCTTTGAAGAACTCAAGGAAGCGCGGCACGATACCGCCCGCAATAAACACGCCGCCAAATGTCCCGAGATTGAGCGCCAGATTGCCGCCAAAACGGCCCATAATGACGCAAAACAGCGACAATGCGCGGCGGCAATCGGTGCAGCTGTCAGCCAGCGCGCGTTCGGTAATATCTTTTGGCTTGAGATTTTCTGGCAGGCGGTTGTCAGCTTTCACAATTGCGCGATACAAATTCACCAGCCCAGGGCCAGAAAGCACGCGCTCCGCCGAAACATGACCAATTTCCGCACGCAATATTTCGAGGATAATGGCCTCTTCTTCACTATTCGGCGCAAAATCAACGTGACCGCCTTCGCCTGGCAAGCTTACCCAACGCTTATCGACATGGACCAGATGCGCAACCCCAAGCCCCGTTCCGGCACCGTAAACCGCAATAGGCTTACCTTCGACCGGTTCTGCGCCACCAAACTGAATCAGATGCTCTTTTTTCAGCATCGGGATCGCCATCGATACAGCGGTAAAATCGTTAATAATTTCCAGATGGCTAAAACCGAGATTCTTTTTCATTTCGGCAATTGAGAACGCCCAGGTATGGTTGGTCATCGCCACCCAGTCACCGGTAATTGGGCAAGCGATGGCAATACAGCCGTCTTTCACCTCGACCTTATGTTCTTCAAGATAAACGCGAATGACCGCTTCGAGGCTGGGGTAATCAAGCCCTGAATAGGTCTTAGCCTGCGAGATTTCACCACTGGCAATATCACACAGAGCAAGACGTGCGTTGGTGCCGCCCTCATCACCGACTAATGCATACTTTGTCATTCTTCAACTGCTCCGCTAAAGTCAAAATAATTCTTTCTCACACTGTAAATACCTGGGGGCATAACAACAACGCCGGAAAGGCAGGCTCCCTGTAAATATCGATCTGGGTCACA
```

```
>lcl|ORF7
MTKYALVGDEGGTNARLALCDIASGEISQAKTYSGLDYPSLEAVIRVYLE
EHKVEVKDGCIAIACPITGDWVAMTNHTWAFSIAEMKKNLGFSHLEIIND
FTAVSMAIPMLKKEHLIQFGGAEPVEGKPIAVYGAGTGLGVAHLVHVDKR
WVSLPGEGGHVDFAPNSEEEAIILEILRAEIGHVSAERVLSGPGLVNLYR
AIVKADNRLPENLKPKDITERALADSCTDCRRALSLFCVIMGRFGGNLAL
NLGTFGGVFIAGGIVPRFLEFFKASGFRAAFEDKGRFKEYVHDIPVYLIV
HDNPGLLGSGAHLRQTLGHIL
```

1. Would this mutation have any effect on how the protein works? What are the factors we should consider before making this prediction?
	1. A changes to a T
	2. If the amino acid changes, and if the change causes a significant structure change, if the change is within the ORF at all, if the mutation is late in the amino acid sequence
	3. Run ORF finder -> BLAST -> `Alignment` Tab
	4. 10th amino acid: V changes to a E
2. It still codes for glucokinase
	1. However, in the UniProt feature viewer, click on the magnifying AA icon
	2. We see that the 10th amino acid is inside the ATP binding site
	3. Its mutation will likely disrupt ATP binding, bad!

``` python
seq1 = 'MTKYALVGDVGGTNARLALCDIASGEISQAKTYSGLDYPSLEAVIRVYLEEHKVEVKDGCIAIACPITGDWVAMTNHTWAFSIAEMKKNLGFSHLEIINDFTAVSMAIPMLKKEHLIQFGGAEPVEGKPIAVYGAGTGLGVAHLVHVDKRWVSLPGEGGHVDFAPNSEEEAIILEILRAEIGHVSAERVLSGPGLVNLYRAIVKADNRLPENLKPKDITERALADSCTDCRRALSLFCVIMGRFGGNLALNLGTFGGVFIAGGIVPRFLEFFKASGFRAAFEDKGRFKEYVHDIPVYLIVHDNPGLLGSGAHLRQTLGHIL'
seq2 = 'MTKYALVGDEGGTNARLALCDIASGEISQAKTYSGLDYPSLEAVIRVYLEEHKVEVKDGCIAIACPITGDWVAMTNHTWAFSIAEMKKNLGFSHLEIINDFTAVSMAIPMLKKEHLIQFGGAEPVEGKPIAVYGAGTGLGVAHLVHVDKRWVSLPGEGGHVDFAPNSEEEAIILEILRAEIGHVSAERVLSGPGLVNLYRAIVKADNRLPENLKPKDITERALADSCTDCRRALSLFCVIMGRFGGNLALNLGTFGGVFIAGGIVPRFLEFFKASGFRAAFEDKGRFKEYVHDIPVYLIVHDNPGLLGSGAHLRQTLGHIL'
for i in range(len(seq1)):
    here = seq2[i]
    if seq1[i] == here:
        continue
    else:
        print(i,here)
```


lecture 3 MIT
5 multiplication methods
bioinform lecture 1 this week, rewatch lecture video
read 2 papers polyphen and sift
finish writing
do math problems, read bio textbook, figure out keywords



