# Data Integration into Genome Scale Models

[0. VU Introduction to Systems Biology](0.%20VU%20Introduction%20to%20Systems%20Biology.md)

## Data types

Metabolomics, proteomics, transcriptomics, fluxomics.

![](Pasted%20image%2020241017144829.png)

## Challenge in data integration

The flux $v$ of a reaction depends on:
- the kinetic properties of the enzyme
- the concentration of the enzyme
- the concentration of the reactants

We cannot get fluxes directly from enzyme concentration. Enzyme kinetics information is also necessary.

An example reaction $S\leftrightarrow P$ 
$$v=k_{cat}[enzyme]\frac{\frac{S}{K_{S}}}{1+\frac{S}{K_{S}}+\frac{P}{K_{P}}}\Big(1-\frac{P}{S K_{eq}}\Big)$$

To get exact flux values, we need:
- the exact rate equation (hard to measure)
- the kinetic parameters of enzymes, the $K$s (hard to measure)
- the reactants concentration, $S$ and $P$ (easy to measure)
- the enzyme concentration, $[enzyme]$ (easy to measure)

## Integration of proteomics data in FBA

How to add proteomics data to [FBA](5.%20VU%20ISB%20Flux%20Balance%20Analysis.md)?
Proteomics data give protein concentration.
Goal: 
- decrease the range flux bounds, giving more confidence in estimated flux
- predict flux from proteomics data
### Ideas

Add it to the **flux bound**?
- If a protein is not detected, its associated flux is also 0 (upper lower bounds are 0).or remove the reactions that have protein concentration of 0. This reduces the stoichiometric matrix, S'.
- Nutrients that are not in the growth medium can also be removed from the network.

Add it to the **objective**?
- Making the objective favor the measured proteins, forcing agreement with biological truth.
- Add a $\sum{v_i\alpha_i}$ to the objective, such as maximizing growth rate. $v_i$ is the flux of protein $i$. $\alpha_i$ is the weight of protein $i$. The higher the measured protein concentration, the higher weight.
- Does not work very well.

Add to **both**?
- A working solution is recognizing that the amino acids produced in the network are used by the cell to make new enzymes, whose concentrations and composition we know. This gives new constraints to the model.

Add **something new**?
- Set the input flux, output flux, and growth rate to measured values. These are easier to measure.
- Compare the predicted flux distribution to the data.