# State Space

[0. VU Introduction to Systems Biology](0.%20VU%20Introduction%20to%20Systems%20Biology.md)

## High throughput experiments

- Genome: sequence of bases
- Transcriptome; number of each transcript
- Proteome: separate protein by m/z ratio. Break the protein into peptides. Peptides measured using mass spectrometer
- Metabolome: LCMS, identify using metabolite databases

## Metabolic network

### How do we measure the properties of an enzyme

- Lower the concentration of all reactants below the affinity of every enzyme
	- By lysing the cell and diluting the biomolecules
- Add specific reactants at a higher concentration
	- So they react with their enzymes
- Measure the change in metabolites
- Create mutants and redo the experiment
	- To know which enzymes are important
- Use this curated knowledge, together with data from high throughput experiment to understand the cell biology
	- Like genomics and proteomics data

### Workflow

![](Pasted%20image%2020240912135130.png)
- Genome > genes > proteins, which are enzymes
- Enzymes are part of the metabolic network
- Model the metabolic network
	- Integrate with proteomics information to know what enzymes are really there.
	- Refine the model
- Use the model to calculate flux

## Conway's Game of Life

[Game](https://sciencedemos.org.uk/conway_game_of_life.php)

Overpopulation or underpopulation causes cells to die. Right population leads to reproduction.

Cells can end up all dead, or steady state, or oscillating.

The rules are simple, but predicting the outcome is still difficult.

Halting problem
- Impossible to have a generalized ML algorithm to predict if the system reaches steady state for any system with any rules given the starting state.

## State space

- State space: a set of variables that describe it
	- A geometric space in which the axis are the state variables
- For the game of life, the variables are the status of each cell. The state is the two dimensional grid.
- **Evolution law**: At $t$, system at state $s$, the state at $s+dt$ is $$f:s(t)\rightarrow s(t+dt)$$
	- For the game of life, the evolution laws are the rules of the game.
- **Dynamics**: the outcome of many subsequent state transitions, applying the evolution law over and over.
	- Hard to predict.
- **Steady state**: $$s(t+dt)=f(s(t))=s(t)$$
	- The state no longer changes.

### Genome

If we consider the genome as a system. Its states would be evolution of the genome. Its evolution law would be natural selection.

### Transcriptome

The expression levels are the variables. The expression matrices at time points are the states. The evolution laws are gene regulation rules

### Proteome

- State space: 3D structure of the protein, and its charges. (e.g. 3D coordinates of all amino acids).
- Evolution law: basic principles known (non covalent bonding, hydrophobicity, entropy), but extremely hard to compute. 
- Dynamics: protein folding as function of time.
- Stationary state: folded protein.

### Metabolome

- State space: The concentration of metabolites and proteins.
- Evolution law: Diffusion, collisions, kinetics, mass balance, regulation. Hard mechanistic biology.
- Steady state: the concentration of the metabolites do not change.

### Very hard to model biological systems

Biology is between orderly and chaotic.
- Orderly side: Newtonian physics, relativity
- Chaos side: ideal gas law, statistical physics
- Biology: in the middle, influenced by both orderly and chaotic rules.

## Systems biology and bioinformatics at large

How systems biology and bioinformatics find answers.
- Systems biology at the extreme
	- Completely mechanistic
	- Can't know all the mechanisms
- Bioinformatics at the extreme
	- Only statistics (correlations)
	- Does not learn the evolution laws

