# Dynamic Programming

[Course link](https://www.youtube.com/watch?v=AuXp5IZNq70&list=PLypiXJdtIca6dEYlNoZJwBaz__CdsaoKJ&index=2&pp=iAQB)
[Slides](http://stellar.mit.edu/S/course/6/fa10/6.047/courseMaterial/topics/topic3/lectureNotes/Lecture02_-_Dynamic_Programming/Lecture02_-_Dynamic_Programming.pdf)
[0. MIT CompBio Manolis Kellis](0.%20MIT%20CompBio%20Manolis%20Kellis.md)

## The Evolution

- All creatures share the common ancestry

## Comparative Genomics

- By comparing the genomes of different species of animals, common regions can be identified
- The closer the evolutional relationship (different species of fruit flies), the more similar the genome: same order, same orientation
- The conserved regions have functional importance
- Only 1.5% of human genome codes for proteins, while the rest 98.5% are found to be non-conserved
- Comparative genomics: *align* genomes to identify *conserved* regions and *discover* functions
- Develop methods for estimating the level of constraint of evolution:
	- Count the number of **edit** operations, number of **substitutions** and **gaps**
	- Estimate the number of **mutations** (including estimate of back-mutations)
	- Incorporate information about neighborhood: conservation 'windows'
	- Estimate the probability of a **constrained "hidden state"**: with HMMs
		- The hidden state is the cause of the observed state. 
			- For example, whether the ground is wet is an observed state. Whether it rained is a hidden state. 
			- A set of data points is an observed state. The underlying distribution is the hidden state.
		- Observed state: amount of conservation
		- Hidden state: probability of being conserved, which is higher in conserved regions
		- The hidden state tells you if the observed conservation is caused by chance or true conservation through evolution
	- Use phylogeny to estimate tree mutation rate, or 'rejected substitutions'

## Genome Alignments

- Genomes must be aligned to study evolutionary signatures for diverse functions.

> [!note] Evolutionary Signatures
> Protein coding genes:
> - Codon substitution frequencies
> - Reading frame conservation
>   
>  RNA structures
>  - Compensatory changes
>  - Silent G-U substitutions
> 
> [microRNAs](1.%20MIT%20CompBio%20-%20Introduction.md#^ae0496)
> - Shape of conservation profile
> - Structural features: loops, pairs
> - Relationship with 3'UTR motifs
> 
> Regulatory [motifs](1.%20MIT%20CompBio%20-%20Introduction.md#^6a1795)
> - Mutations preserve consensus
> - Increased branch length score
> - Genome-wide conservations

> [!note] Two Types of Alignment
> 
> - Global alignment: an attempt to align every element in a genetic strand, most useful when the genetic strands under consideration are of roughly equal size and expected to be **related along their entire length**. Note that in a variation of global alignment called semi-global alignment, **gaps** are permitted at the beginning and/or end so that the entire length of one sequence is aligned to a portion of the other.
> - Local alignment: an attempt to align regions of sequences that contain **similar sequence motifs** within a larger context.

### Formulate the problem

> What should be considered when we align two sequences?

- Genomes change over-time: 
	- mutations: edits
	- deletions, insertions: gaps
- Problem formulation:
	- Define set of evolutionary operations: should be symmetric operations that allow time reversibility
		- For example, insertion is reversible by a deletion
	- Define optimality criterion: minimum number, minimum cost
		- **Occam's razor**: choose the simplest model that explains the data, also known as the maximum parsimony (minimize spending) principle
	- Design algorithm that achieves that optimality
		- Tractability of solution depends on assumptions in the formulation
			- Tractability: the problem complexity is manageable
		- Trade-off between being biologically correct and computationally tractable; can be both

> [!info] Time reversibility is a design principle
> Time reversibility is a design principle. For example, methylated CpG may incorrectly pair with a TpG or CpA during DNA replication due to deamination of the methylated C, but the reverse operation cannot occur.

> [!info] Other optimality criterions
> The optimality criterion does not have to be choosing the simplest model. We could choose a probabilistic method, for example using Hidden Markov Models (HMMs), that would assign a probability measure over the space of possible **event paths** and use other methods for evaluating alignments (e.g. [Bayesian methods](MLE%20vs.%20Bayesian%20Parameter%20Estimation.md)).
> 
> Note that in choosing the simplest model, the underlying idea is that mutations are a rare event, and in minimizing the number of mutation events, we *maximizes its likelihood*. While Bayesian views of probability underlies the Bayesian optimality criterion. See [MLE vs. Bayesian Parameter Estimation](MLE%20vs.%20Bayesian%20Parameter%20Estimation.md).
 
### Formulation 1: Longest common substring

- Given 2 strings S1 and S2, find the longest substring without gaps
- Two answers ![](Pasted%20image%2020240422163941.png)
- This won't account for insertions and gaps
- See also [Longest Common Substring Problem](Longest%20Common%20Substring%20Problem.md)

### Formulation 2: Longest common subsequence

- Allow gaps
- More nucleotides identified as conserved ![](Pasted%20image%2020240422164400.png)
	- Insertion of T
	- Deletion of C and A
	- Mutation of C to G
- Number of changes: the number of operations
- Uniform scoring function: each operation has the same cost

### Formulation 3: Sequence alignment

- Fixed penalty for gaps: each insertion or deletion operation has the same cost.
- Varied penalty for edits: takes into consideration the fact that some edit operations are less likely than others
	- Transition: transition within purines (A and G) and pyrimidines (C and T) are more likely
	- Transversions: purine - pyrimidine changes. Less likely since purines are large [nucleotides](1.%20MIT%20CompBio%20-%20Introduction.md#^4f79e6), and pyrimidines are small nucleotides
	- Polymerases confuses A with G and C with T more often
- No change is rewarded with 1; mismatch A and G or C and T is penalized with $-\frac{1}{2}$; others are penalized with -1

### Formulation 4: Varying gap cost models

- Linear gap penalty: same penalty for all gaps
- Affine gap penalty:
	- Big initial cost for starting a gap
	- Small incremental cost for each additional nucleotide
- General gap penalty:
	- Any cost function
	- No longer computable using the same model
- Frame-aware gap penalty
	- Gap sizes that are not multiples of 3 will disrupt the reading frame for translation
	- They are more penalized
- Seek duplicated regions, rearrangements, ...

> Which one to choose depends on the required speed and biological problem. For example, linear gap is the fastest, but not very biologically relevant. Affine is slower, but reflects more biological truth.

### How many alignments?

![](Pasted%20image%2020240508190156.png)
- When aligning two sequences of length m and n, the **longest** non-boring alignment has m+n entries, such that a gap is not aligned to another gap.
- Alignment is equivalent to gap placement.
	- There are (n+m choose n) ways to place gaps for S1
	- At each position, choose whether to place gap or not
	- This is an exponential problem; impossible to enumerate every alignment and score them
- [Dynamic programming](Dynamic%20Programming.md): solving an exponential problem with polynomial time

## Dynamic Programming for Sequence Alignment

> See [Dynamic Programming](Dynamic%20Programming.md)

### Sequence alignment is suitable for dynamic programming

- The sequence alignment problem has optimal substructure, meaning the optimal alignment solution is made of optimal solutions to subproblems.
- It also has overlapping subproblems. The alignment of longer sequences is based on the result of aligning shorter sequences.

### Matrix parameterization

- S1 is placed on the columns, S2 is placed on the rows.
- A gap is added before S1 and S2 to account for zero overlaps between the sequences
	- More specifically, they are the gap scores for aligning S1 with all gaps and S2 with all gaps
- The top left is 0.
- There are 4 cases needed to be considered:
	- S1 has a gap at the current alignment position
	- S2 has a gap at the current alignment position
	- A mismatch happens at the current alignment position
		- Different scores can be given to different mismatches, see [Formulation 3 Sequence alignment](2.%20MIT%20CompBio%20-%20Dynamic%20Programming.md#Formulation%203%20Sequence%20alignment)
	- A match happens at the current alignment position

### Traversing the matrix

- By row or by column
- To calculate the score, take the max(left+current, top+current, and diagonal+current) score
	- From the left: extend S2 by a gap, extend S1 by a character
	- From the top: extend S1 by a gap, extend S2 by a character
	- From the diagonal: extend S1 and S2 by a character
- Another matrix is generated to record in which direction does the maximum come from, aka. a pointer. When there are >1 equal scores, record all of them.

### Backtrack

- After all the scores are calculated, travel from the bottom right to the top left
- When there are multiple pointers, go separate ways
- When the score comes from a score diagonally, add both characters
- When the score comes from a score on the left: add a character from S1 and a gap for S2
- When the score comes from a score on the top: add a character from S2 and a gap for S1

### Example

Simplified example using linear gap and edit penalty
Gap score: -2
Match: 1
Mismatch: -1

|     | Gap                                           | A                                              | C                                              | G                                              | T                                              | C                                             | A                                             | T                                             | C                                             | A                                             |
| --- | --------------------------------------------- | ---------------------------------------------- | ---------------------------------------------- | ---------------------------------------------- | ---------------------------------------------- | --------------------------------------------- | --------------------------------------------- | --------------------------------------------- | --------------------------------------------- | --------------------------------------------- |
| Gap | <mark style="background: #FF5582A6;">0</mark> | -2                                             | -4                                             | -6                                             | -8                                             | -10                                           | -12                                           | -14                                           | -16                                           | -18                                           |
| G   | -2                                            | <mark style="background: #FF5582A6;">-1</mark> | <mark style="background: #BBFABBA6;">-2</mark> | <mark style="background: #FFB8EBA6;">-1</mark> | -2                                             | -3                                            | -4                                            | -5                                            | -6                                            | -7                                            |
| T   | -4                                            | -2                                             | <mark style="background: #ADCCFFA6;">-2</mark> | <mark style="background: #BBFABBA6;">-2</mark> | <mark style="background: #FFB8EBA6;">0</mark>  | -1                                            | -2                                            | -1                                            | -2                                            | -3                                            |
| G   | -6                                            | -3                                             | -3                                             | <mark style="background: #ADCCFFA6;">-1</mark> | <mark style="background: #FFB8EBA6;">-1</mark> | -1                                            | -2                                            | -2                                            | -2                                            | -3                                            |
| T   | -8                                            | -4                                             | -4                                             | -2                                             | <mark style="background: #FF5582A6;">0</mark>  | -1                                            | -2                                            | -1                                            | -2                                            | -3                                            |
| C   | -10                                           | -5                                             | -3                                             | -3                                             | -1                                             | <mark style="background: #FF5582A6;">1</mark> | 0                                             | -1                                            | 0                                             | -1                                            |
| A   | -12                                           | -4                                             | -4                                             | -4                                             | -2                                             | 0                                             | <mark style="background: #FF5582A6;">2</mark> | <mark style="background: #FF5582A6;">1</mark> | <mark style="background: #FF5582A6;">0</mark> | <mark style="background: #FF5582A6;">1</mark> |

The pink fork:

```
ACGT--CATCA
G--TGTCA---
```

The blue fork:

```
ACGTCATCA
GTGTCA---
```

The green fork:

```
ACG-TCATCA
G-TGTCA---
```

Overall the blue fork involves the least cost.

## Improvements on the algorithm

### Bounded Dynamic Programming

- Most of the time, the matrix is traversed along the diagonal, not deviating too far from it.
- Therefore, if we only compute the scores along a diagonal region, the speed of the algorithm can be improved.
- Time: O(k\*m); Space: O(k\*m), k is the radius explored.
- Although this no longer guarantees finding the optimal alignment, it still works well in practice.

### Hirschberg Algorithm

#### Linear space alignment

- Another way to speed up computation is to reduce the space requirement for storing the matrix.
- The score matrix can be computed column by column, and when the next column is computed, the previous column can be discarded to save space.

#### Traceback

- By removing the previous columns, we can't traceback in the old way.
- Beginning at the top left and bottom right, compute the scores toward the center. At the center column, sum the scores. The maximum score lies on the optimum path;
- Then use the pointers going into and leaving the center column, get two sub-matrices that are each part of the left and right matrices;
- Recurse the above two steps to find the maximum path.
- The run time is $O(mn)$; the space requirement for computation is $O(m)$; the space requirement for storing the optimal alignment is $O(m+n)$

![](Pasted%20image%2020240513181150.png)