# Visualizing and Understanding Gene Expression Regulation in Single Cells

[0. VU QSCB](Biology/VU%20Quantitative%20Single%20Cell%20Biology/0.%20VU%20QSCB.md)
> Evelina

## Single cell biology

- Measure characteristics and behaviors of individual cells
	- growth rate
	- mRNA
	- protein
	- cell size
- how differences in _cell states_ lead to distinct _cell fates_
- collective behavior emerging from single cells

## smFISH

- Single molecular RNA detection
- Get mRNA counts

## Background

### More than 200 cell types in the human body

- All differentiated from stem cells
- Where does their diversity come from?
- Cell type is dependent on which genes are activated

### The central dogma of biology

- DNA-RNA-protein
- Several mechanisms controls mRNA levels and in turn protein levels, such as transcription, splicing, localization, mRNA decay
- Measuring mRNA levels can help elucidate gene expression regulation mechanisms

## Bulk mRNA analysis methods

- Low to mid-throughput techniques (1-10 mRNA species)
	- Northern blot
	- Reverse transcription PCR (RT-PCR) and qRT-PCR
- High-throughput techniques (>10 mRNA species):
	- DNA microarray
	- RNA-Sequencing
		- Nanopore sequencing

### Northern blot

- RNA isolation
- Electrophoresis
- Separation by size
- Transfer on membrane
- Labeled probes
- Visualization of bands on film or digital imager

### RT-PCR and qRT-PCR

- RNA extraction
- Reverse transcription to purified RNA to cDNA
- Store cDNA in -20 C and amplify using PCR
- Measure mRNA level with Ct values - more mRNA leads to less Ct value

### DNA microarray

- Cell type
- Culture
- RNA isolation
- Reverse transcription and fluorescent tagging
- Hybridization onto microarray
- On the array: present in normal cell, present in pathological cell, not present, and present in both cells

### RNA sequencing

- Total RNA extraction
- Reverse transcription: RNA to cDNA
- Attach to adapters then attach to flowcell
- PCR expansion on flowcell
- Number of mRNAs in mRNA clusters correspond to mRNA level
- Sequencing by synthesis: as fluorescently tagged nucleotides are added, the detector records the added nucleotide
- Advantages:
	- Massively parallel sequencing technology
	- High throughput
	- Scalability and speed
	- Possible applications:
		- Quantify mRNAs for gene expression analysis
		- Discover novel RNA variants and splice sites
		- Analyze regulatory factors (e.g. RNA-protein interactions)
		- Sequence cancer (or any disease) samples to identify rare somatic variants, tumor sub-clones, etc.

### Nanopore sequencing

- RNA extraction
- No need for RNA -> DNA conversion
- The RNA/DNA molecule is unwound by the helicase protein and a single strand is translocated through the pore to the + side of the membrane
	- RNA passes through pores on a resistive membrane that forms a stable electric field
- Nucleotides disrupts the electric field, and the disruption is characteristic of the nucleotide
- Can sequence long mRNAs
- Modifications (e.g. methylation) can be detected
- ML algorithms interpret the signal patterns and convert them to a nucleotide sequence (base calling)
- Advantages:
	- High throughput
	- Scalability and speed (results in real-time)
	- Long-read sequencing

## Why study gene expression in single cells

- Bulk RNA-seq does not contain single cell and spatial information
	- Measures the average cell

## Single cell gene expression analysis methods

| RNA expression                                                                               | Protein expression                                                                                        | Advantages/Disadvantages                                                                                                                                                                           |
| -------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Single cell RNA sequencing                                                                   | Single cell proteomics (mass spectrometry)                                                                | High-throughput  <br>- No prior information on target genes is required  <br>- Not (yet) sensitive for low-copy number mRNAs/proteins (tens of  <br>molecules)  <br>- Spatial information is lost  |
| scRNA imaging<br>- RNA fluorescent in situ hybridization<br>- RNA reporters for live imaging | Single cell protein imaging<br>e.g. tagging with fluorescent proteins (YFP reporter), immuno-fluorescence | - Sensitive to detect low-  <br>abundance mRNAs  <br>- Spatial information is preserved  <br>- Low-throughput (however,  <br>recent developments improved the number of mRNA that can be targeted) |
|                                                                                              |                                                                                                           |                                                                                                                                                                                                    |

### Single cell separation methods

- Pipette
- Microscope  + capillary pipette
- FACS: cell sorting based on laser and multispectral detector + electronics
- LCM
- Microfluidics
- Separated cells used for downstream scRNA or proteomics

## Single molecule RNA in situ hybridization (smFISH)

- Contains shape and spatial information
- Gene expression level
- mRNA localization
- protein level

### How to perform mRNA imaging to study mRNA distribution

> in the 80's: in situ hybridization with radioactive probes

1. Cellular fixation (paraformaldehyde): to preserve cell shape and mRNAs
2. In situ hybridization: with isotopically labeled probes
3. Detection: photographic emulsion
	1. Emulsion contains Ag+ ions
	2. Radiation activates Ag+ ions
	3. Ions converted to Ag metal
	4. Ag metal can be detected with electron microscopy

#### Application

- First evidence of mRNA localization by in situ hybridization
- Autoradiograph of chicken fibroblast cells hybridized in situ with ${}^3$H labeled probes to $\beta$-actin mRNAs encoding the cytoskeletal protein Actin

#### Major limitations

- To detect the signal the photographic emulsion is exposed for weeks/months
- Radioactive (${}^3$H) decay causes spreading of the signal onto the photographic emulsion
- Thus, localization of the point source is not accurate ($\mu m$ away). Low resolution.

###  A new approach: using fluorescent probes

- Label probes with fluorophores, not fluorescent proteins, are used to label mRNAs. 3-4 mRNAs
- Fluorescence allows signal isolation from the focal plane (sharp ~ 200-300 nm)
- Advantages:
	- Higher resolution
	- Increased speed & safety
	- Multi-color
	- Quantitative
	- Live-cell imaging

#### Application

- Simultaneous hybridization to both viral RNA and DNA in denatured samples.  
- Spots of yellow fluorescence on each sister chromatid of chromosome 1 of human lymphoma cells indicate the localization of the integrated EBV genomes

### smFISH

- DNA oligos of about 20 nts bound to fluorophores attached to mRNA
- 3D Gaussian fitting on the signal
- image analysis for localized mRNA:
	- cell segmentation
	- 3D Gaussian spots fitting
	- mRNA counting

#### Application

- Fungal spatiotemporal gene regulation across scales

#### Spots localization by wide-field microscopy, image sectioning, and 3D Gaussian fitting

- For each cell, get multiple images on the z axis
- If just have 1 fluorophore probe bound to mRNA, cannot detect the fluorescence above cellular background.
- Instead have 40-50 probes per mRNA
- Because of the resolution, the signals from the probes are blended into one spot
- In microscopy using higher resolution, they can be visualized

#### smFISH revealed that RNA transcription can be Poissonian or Bursty

- **Poissonian**: In which mRNAs are synthesized in random, uncorrelated events, with a probability that is uniform over time
- **Bursty**: In which mRNA is produced in episodes of high transcriptional activity (bursts) followed by long periods of inactivity
- Fast switching of transcription kinetics
	- Unimodal mRNA distribution
- Slow switching of transcription kinetics
	- Bimodal mRNA distribution
- Gene expression activity can be described by a 2-state process: $Off{\underset{k_{off}}{\overset{k_{on}}{\leftrightarrow}}}On{\overset{k_{TX}}{\rightarrow}}mRNA{\overset{k_{d}}{\rightarrow}}\Phi$
- More modes of transcription are being discovered

## Molecular Regulation of Transcription

![](Media/25548.png)
- Transcription activation is rate limited by the binding of sequence-specific factors to DNA proximal to the gene transcription start site, followed by the assembly of the RNA polymerase II and initiation and elongation of the nascent transcript
- genes can form stable complexes with distal enhancers: as long as the complex does not break transcription keeps happening
- enhancers in close proximity to a gene promoter might increase the local concentration of factors needed for activation

## RNA synthesis is a significant source of molecular variability

- Most of the noise that causes cell variability is from transcription
