# Gene Expression Yeast GAL Cluster

[0. VU QSCB](Biology/VU%20Quantitative%20Single%20Cell%20Biology/0.%20VU%20QSCB.md)
> Evelina

---

## Gene expression noise

- Transcription is a main contributor to noise and cell-to-cell heterogeneity

## Yeast gene expression regulation

- 16 chromosomes
- GAL genes are cluster together in the genome and **co-regulated**
- All genes need to be activated for the pathway to activate
- GAL1 and GAL10 share the same promoter

![](Media/Pasted%20image%2020250603110809.png)

### The coordinated expression of the GAL gene cluster

- Galactose is metabolized to glucose 6-phosphate by the enzymes of the Leloir pathway.
- Many of the enzymes of the glycolytic pathway are so specific for glucose that other sugars, even other hexoses (like galactose), are not processed.
- To overcome this problem, there are number of short pathways which convert other common sugars (e.g. galactose and fructose) into glycolytic intermediates.
- In Saccharomyces cerevisiae, the five enzymes required for this conversion are provided by five proteins – Gal1p, Gal7p, Gal10p, Pgm1p and Pgm2p.
	- Gal1p, Gal7p, Gal10p are encoded by three genes found on the same chromosome.
- The control of the expression of the genes encoding the enzymes of the Leloir pathway (the GAL genes) was one of the first eukaryotic gene expression systems to be studied in great detail.
- The GAL system still represents a genetically and biochemically amenable system to elucidate the mechanisms controlling gene expression that extend far beyond the regulation of sugar metabolism in yeast.  

## Chromatin regulation

- Chromatin needs to be first unwinded for transcription
- In eukaryotes, DNA is organized into chromosomes. Each chromosome contains one very long DNA molecule packed around nucleosomes and forming the chromatin fiber.
- Each nucleosome is composed of two loops of DNA, approximately ~147 bp long, wrapped around 8 central histone proteins. A short segment of linker DNA (~20-60 bp) is found between each nucleosome.
- The nucleosomes arrange into coils, and in preparation for nuclear division, the coiled structure coils up again, to form a supercoiled chromosome.
- Packaging of DNA is necessary: in a single human cell there are approximately 2m of DNA distributed between 46 chromosomes, yet the average diameter of each cell is only 10–30 μm (~10’000x compaction during metaphase!).
- Nucleosomes are important for the safe storage of DNA and in the regulation of transcription.
- The chromatin must be in an open structure and nucleosome octamers must be removed from the gene promoter before RNA polymerase can bind. 
- Regulation of eukaryotic transcription is influenced by the chromatin structure
- A significant difference between the transcription of eukaryotic and prokaryotic RNAs is that in bacteria, transcription takes place on a ‘naked’ circular DNA template, whereas in eukaryotes, transcription takes place on a chromatin template, which is composed of DNA and nucleosome proteins.

## Regulation at promoter

- Proper regulation of transcription is essential for cellular survival   through the establishment and maintenance of homeostasis within a particular environmental and developmental context.
- To balance transcription on a genome-wide scale, many layers of   regulation have evolved and are utilized, including **cis regulatory** DNA sequences, such as promoters and enhancers, that recruit **trans-acting**, DNA-binding transcription factors.
- The **accessibility** of cis regulatory sequences, and the corresponding transcription factors, is influenced by the formation of differential _chromatin structures_ resulting from changes in the location, modification, and composition of histones and nucleosome complexes
	- More open chromatin allows easier binding of TFs to promoters

### Transcription factors

- Transcription factors are proteins that influence the probability of RNA polymerase to transcribe a given gene in trans.
- There are two main types
	- General transcription factors
		- Required for the binding of the RNA polymerase to the core promoter and its progression to the elongation stage
		- Are necessary for basal transcription
	- Regulatory transcription factors
		- Recognize cis regulatory elements located near the core promoter of specific genes. The binding sites are known as response elements, control elements, regulatory elements, or cis-acting elements
		- The binding of these proteins to these elements, affects the transcription rate of an associated gene
- A regulatory protein that increases the rate of transcription is termed an activator
- Promoter contain various binding sites >> one is ‘enhancer’
- A regulatory protein that decreases the rate of transcription is termed a repressor

### Transcription regulation

- To balance transcription on a genome-wide scale, myriad layers of regulation have evolved and are utilized, including cis regulatory DNA sequences, such as promoters and enhancers, that recruit trans-acting, DNA-binding transcription factors.
- Transcription factors work in either singular or combinatorial fashion to recruit RNA polymerases, ultimately altering the rate of production of RNA to meet cellular demands.
- Most Eukaryotic genes are regulated by many factors. This is known as _combinatorial control_ 
- The accessibility of cis regulatory sequences, and the corresponding transcription factors, is altered by the formation of differential chromatin structures resulting from changes in the location, modification, and composition of histones and nucleosome complexes.

## Coordinated expression of gene clusters

Many co-regulated gene families utilize _spatial positioning_ to coordinate transcription  
• Functionally related genes cluster together allowing coordinated expression via adjacent gene co-regulation.

This favors the coordinated expression of functionally related gene families whose composite members are required in stoichiometric levels by the cell.  
• The best characterized examples of this are represented in the co-regulated gene families, or regulons, that are required for production for instance, of ribosomes, cell cycle regulators or biochemical pathways (e.g. GAL cluster)  
• In S. cerevisiae, spatial proximity of genes allows for adjacent gene co-regulation, whereby the clustered genes are co-regulated via shared regulatory elements functioning at a distance.  
• Many functionally related gene families exhibit this nonrandom distribution of their members throughout the genome as functional clusters. This characteristic is conserved in broadly divergent eukaryotes and coordinates transcriptional co-regulation of these gene families

## Spatial coordination of gene clusters expression

- spatial chromosome organization is not random but structured to express certain genes together.
- _transcription hubs_ identified in the nucleus: regions where transcription factors cluster.

---

## Transcriptional repression of GAL genes

- Two types of repression
- _Specific_ GAL genes repression: Gal80 binds Gal4, preventing transcription
- _Carbon catabolite_ repression:
	- Glucose and Fructose are S. cerevisiae preferred carbon sources
	- When one of these sugars is present, the enzymes required for the utilization of _all_ alternative carbon sources (e.g. galactose, maltose) are synthesized at low rates or not at all.
	- Under high glucose, Snf1 complex is inactive, Mig1 is not phosphorylated, it stays in the nucleus and represses alternative carbon source genes.
	- Under low glucose, Snf1 complex is active, Mig1 is phosphorylated, it exits the nucleus and alternative carbon source genes are not repressed.

## Transcriptional activation of GAL genes

- Gal4p is a DNA-binding transcriptional activator, essential for activation of GAL genes
	- Gal4 protein is a Zinc-finger protein, binding as a dimer to the consensus DNA binding site CGGA(G/C)GACAGTC (C/G)TCCG (UAS)
- The presence and number of DNA sequence elements on GAL promoters confer galactose “inducibility” and control strength and duration of transcription activation
- In a wild-type strain growing in a non-inducing medium (e.g. glucose, raffinose) Gal4p remains bound to the specific sequences referred to as upstream activating sequences (UAS), but it is unable to activate transcription due to the inhibition conferred by Gal80p, the repressor.
- At the presence of galactose, the GAL genes move closer to the nuclear pores to make moving activating/repressing transcription factors in and out of the nucleus easier.

### Bi-directional promoter

![](Media/Pasted%20image%2020250603121514.png)

- RNA imaging: MS2 system, PP7 system allows visualizing mRNA molecules in living cell
	- Modify the DNA sequence to include MS2 and PP7 that are transcribed together with GAL genes
	- MS2 and PP7 form loops that bind to code proteins
	- Code proteins can bind to fluorescence proteins like GFP
	- MS2 and PP7 operate orthogonally
- In wild type, GAL1 and GAL10 are on the same chromosome, and their transcription is coordinated.
- In a modified strain, GAL1 and GAL10 are placed on different chromosomes, and the coordination is lost.

### Summary GAL genes activation and repression

![](Media/Pasted%20image%2020250603121609.png)

## Transcription de-repression

- Cells switch from galactose, glucose, then back to galactose
- When glucose is reintroduced, the cells do not degrade all galactose enzymes
- When glucose is removed and galactose is re-introduced, the cells have a _faster recovery_ of galactose metabolism

## Bistability and bimodality

- When carbon source switch from glucose to galactose, a _subset_ of cells switch to galactose metabolism, some remain in glucose mode (bet hedging)
- When carbon source switch from raffinose to galactose, because raffinose _contains galactose_, the cells are primed for galactose metabolism, all cells convert to galactose metabolism

---

## Summary

- In the budding yeast Saccharomyces cerevisiae, the five enzyme activities are provided by five proteins – Gal1p, Gal7p, Gal10p, Pgm1p, and Pgm2p belong to the Leloir pathway and they convert galactose to Glucose 6P.
- The GAL system still represents a genetically and biochemically amenable system for elucidating mechanisms of gene expression that extend far beyond the regulation of sugar metabolism in a single celled organism.
- The GAL genes are spatially organized in the same cluster.
- GAL genes can switch between different transcriptional states: repressed; de-repressed or active.
- The repression relies on two independent mechanisms to keep the GAL genes shut down when not required.
	- A general Catabolite repression and a specific GAL gene repression.
- Activation relies on the number of binding sites of the Gal4 transcriptional activator.
- Transcriptional memory allows cell to adapt faster to recurrent metabolic transitions to achieve optimal growth.
- Transcription is not only controlled by the current metabolic state of a cell, but also by its previous experience. Thus, single cell methods allowing us to measure gene expression changes over time are powerful tools to elucidate how cells dynamically adapt to environmental changes.