=# Introduction

[0. VU Structural Bioinformatics](Biology/VU%20Structural%20Bioinformatics/0.%20VU%20Structural%20Bioinformatics.md)

## Why study structures

- link mutations to phenotypical changes
- alpha-fold cannot predict everything
	- biased for proteins that can crystallize

## Structural bioinformatics

- Comparing structures
	- Between 1 and many: validate the experimental reliability of the structure
	- Between 1 and 1 or more: how similar the protein folds are
- Sequence -> structure
- Structure -> function
- Structure dynamics
	- structure prediction
	- statistical thermodynamics
	- simulation
	- how structure conformation change affects function
	- view protein structure as an ensemble of structures

## Course material

[Course book](Biology/VU%20Structural%20Bioinformatics/slides/strucbioinf_students_shrunk_ebook.pdf)
3 papers

## Protein structure basics

- primary: sequence of amino acids
- secondary: $\alpha$-helices, $\beta$-sheets, loops
- tertiary: folding into 3d structure
- quaternary: interaction between proteins forming macro-molecular complexes
- Hydrogen bonds 
- Dihedral angles

### Primary structure

![](Pasted%20image%2020250203114659.png)
- Structure of an amino acid:
	- $R$: side chain
	- $NH_2$: amino group
	- $COOH$: carbonyl group
	- $H$: hydrogen atom
	- $C_\alpha$: alpha carbon 
- 20 naturally occurring amino acids
- linked together by **peptide bonds** between the carbonyl carbon of the previous AA and the amide nitrogen of the next AA in the sequence
	- extremely stable
	- ![](Pasted%20image%2020250203114947.png)

#### Amino acids

- [Transcription](Transcription.md) + translation by [Ribosome](Ribosome.md)
- three classes: hydrophobic (aliphatic), charged (positively and negatively), polar (not charged)
	- based on interaction properties with water
	- likes interact well with likes
- Aromatic: contains rings in side chain:
	- Tyr: aromatic and polar
	- Trp: aromatic and hydrophobic
- Polar and charged AAs all have N or O atoms in the side chain.
	- charged are also considered polar
	- charged and polar are both hydrophilic
- Classification in size
	- hydrophobic AAs can be further divided into small (Gly, Ala), medium (Val, Leu, Ile), and large (Met) in size
	- aromatic AAs: Pro, medium; Phe, Trp, large
	- charged: Lys, Arg, large
	- polar: Ser, Thr, large
- Special amino acids:
	- Cysteine: 
		- form disulfide bond with another cysteine
		- post-translational modification
	- Proline  
		- Side group forms ring with backbone
		- Much less flexible
		- often terminates helices or otherwise induces a kink, and proline is used to make a loop containing sharp turns
		- Different phi/psi angles  
	- Glycine  
		- The smallest side chain, only a H atom 
		- Less steric hinderance
		- More flexibility in phi/psi angles  
- **Question**: how can the order of these amino acids determine the shape of the protein
	- _hydrophobic effect_: proteins fold such that hydrophobic AAs are inside, and hydrophilic AAs are outside
	- so for a helix, the part of it that faces _inwards_ should be _hydrophobic_ and that part that faces _outwards_ should be _hydrophilic_ for it to fold properly

### Secondary structures

- formed from **hydrogen bonding** of the backbone

#### Hydrogen bonding

- Hydrogen bonds are caused by atoms with slight negative charges (δ−) being attracted to atoms with slightly positive charges (δ+). 
- In a protein, hydrogens on a nitrogen or oxygen are _positive_, oxygens and nitrogens themselves are _negative_.
- can be formed with:
	- backbone: N-H and C=O
	- side chain: polar and charged side chains (O-H, N-H, S-H, and C=O groups)
	- water
- _very_ energetically favorable

#### Amino acid backbone

- The backbone is hydrophilic (contains O and N)
- Problem: when hydrophobic AAs are folded inside the protein, the hydrophilic backbone is also inside. This is not energetically favorable
- Solution: the charged atoms (O and H) on the backbone hydrogen bond to each other to neutralize the charges
- Backbone hydrogen bonds in the hydrophobic interior can become satisﬁed by **helix** and **strand**

#### Alpha helix

- A chain H-bond with itself: repeated local hydrogen bonding
- Unsatisfied H-bonding capacity at the end of the helix, which are found at the _surface_ of a protein
- Displays a periodicity in the sequence and is easier to predict from the sequence

#### Beta strand

- Beta-strand: a stretch of amino acids with the backbone in an extended configuration, typically 3 to 10 amino acids long. 
- Two or more $\beta$-strands together make up a $\beta$-sheet.
- A strand H-bond with another strand
- Unsatisfied H-bonding capacity at the first and last strands of a sheet (edge strands), which are found at the _surface_ of a protein
- Two directions: 
	- parallel: N to C binds to N to C; distinguished by slanted H-bonding pattern
	- anti-parallel: N to C binds to C to N; linear H-bonding pattern

#### Loops

- Not helices or sheets; no regular pattern of H-bonding
- The H bond donors and acceptors still need to H-bond with:
	- other backbone groups
	- side chains of polar AAs
	- water
- Tends to be more flexible, _more on the surface_, solvent-exposed
	- More likely for indel [Mutation](2.%20VU%20FoB%20-%20Mutation,%20Evolution,%20and%20Sequence%20Alignment.md#Mutation)
	- Contain more gaps in multiple sequence alignment
- **Disordered regions**: very long loops ^fa1a21
	- will not take up a rigid 3D structure in their folded state

#### Other unusual 2nd structures

- amyloid fibrils
- coiled-coils
- disordered proteins
- see textbook p.29

#### Phi & Psi Angles

![](Pasted%20image%2020250203114659.png)

- N-$C_\alpha$ bond and carboxyl-$C_\alpha$ bonds can rotate. When they rotate, the AA chain connected to them also rotate.
- Specific combinations of their rotation angles allow the formation of $\alpha$-helices and $\beta$-sheets, while others can result in clashes within the backbone or between adjacent side chains in the AA chain, particularly if the side chains are large.
- Clashes limit the amount of conformations that the polypeptide may adopt.
- The rotation angle of the N-$C_\alpha$ bond is the $\Phi$ (Phi) angle 
- The rotation angle of the Carboxyl-$C_\alpha$ bond is the $\Psi$ (Psi) angle 
- A scatter plot (_Ramachandran plot_) of the 2 angles can be made, and different 2nd structures would occupy different regions, leaving areas corresponding to infeasible combinations empty

#### The omega torsion angle

![](Pasted%20image%2020250203145224.png)
- The peptide bond between N and Carboxyl
- Has an angle called $\Omega$
- Cannot rotate freely, and it makes the bond lie in a plane
	- the polypeptide chain is a sequence of planes interconnected by rotatable amino-$C_\alpha$ bond and carboxyl-$C_\alpha$ bonds 
	- but in proline it can take two possible angles in a process called _proline isomerisation_
	- it is special as proline's side chain forms a loop with the backbone

#### Secondary structure assignment

- Assign 2nd structure labels for each AA based on protein structure
- Protein structures are typically stored as a set of coordinates for each atom in the structure
- Various features such as phi and psi angles or hydrogen bonding patterns, can be used to assign secondary structure.
- DSSP
	- Assigns H-bonds to pairs of atoms
	- Use H-bonds to infer secondary structure
		- if there are consecutive H-bonds to AAs 4 places ahead, assign $\alpha$-helix
	- Minimum length of 2nd structures is also considered

### Tertiary structure

- Contain 2nd structure or _motifs_
	- Motifs: particular combinations of 2nd structures, can be informative for assessing functions
- How the secondary structures are arranged and folded into a 3d structure (protein topology)
- 4 main fold categories: all-$\alpha$, all-$\beta$, $\alpha$/$\beta$, and $\alpha+\beta$
	- can be considered as motifs
- Databases:
	- Topology (SCOP)  
	- Architecture (CATH)

#### Hydrophobic core

- Proteins are always in water, except trans-membrane proteins
- Hydrophilic AAs outside, hydrophobic AAs inside to avoid contact with water: **hydrophobic collapse**
- Hydrophobic effect is the **main driving force** of protein folding
- Folded protein tends to have regular 2nd structures (helices and strands) at the core to satisfy the H-bonding potentials the hydrophilic backbones at the core [Amino acid backbone](#Amino%20acid%20backbone)

#### Protein domains

- conserved regions that are self-folding, evolutionarily conserved subunits of structure
- specific molecular function
- _recurrently_ appear in different proteins
- domains may be _linked_ by a small linker or large disordered regions (long [Loops](#Loops))
- structure prediction is most effective at the **domain level**

### Quaternary structure

- Protein complex formation
- Example: Bovine f1-ATPase: different protein work together to drive the flagella
- [7. VU BMBN Multimeric Enzyme Kinetics](Biology/VU%20Basic%20Models%20of%20Biological%20Networks/7.%20VU%20BMBN%20Multimeric%20Enzyme%20Kinetics.md)

## Structure & function

- Classic picture: sequence to structure to function
	- overly simplistic
	- no single structure but dynamic
	- no single function but multiple functions
- Or: once we have structure, function will follow

### Functions

- Protein binding sites (enzymes)
- Protein interaction surface
- How to recognize the functional site
	- Given structure?
	- Given sequence?
- Functional sites often located on 'structural rule' exceptions

## Protein-protein interactions (PPI)

- some may be functional, some may not be

## What does bioinformatics do

- What can we do with these structures
- Where do the data come from
- Predict how proteins fold

### PDB

- [Protein DataBank](www.ebi.ac.uk/pdbe/)
	- X-ray structures  
	- NMR structures  
	- cryo-electron microscopy  
- Experimental bias:  
	- which proteins can we _purify_, _crystallize_ or _stabilize_  
		- all non-trivial steps
		- trans-membrane proteins are hard to express, purify, crystallize, and stabilize in solution
	- Sequence bias
	- Some proteins/families much more studied than others!  
	- 200.000 structures  
	- ~10.000 clusters  

### AlphaFold DB

- https://alphafold.ebi.ac.uk/
- Trained on structures from Protein DataBank  
- Also trained on sequence data (details in later lecture)
- Shares biases with PDB!

### Prediction approaches

- Knowledge based:  
	- Data derived  
	- Mined over extended  
	- experimental datasets  
- Ab initio:
	- From first principles  
	- Based on physical laws

### PDB file format

- 3D coordinates of atoms in space
- summary information about the structure, sequence, and experiment
- Challenges: 
	- many structures, particular those determined by crystallography, only include information about part of the [functional biological assembly](https://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/biological-assemblies). Fortunately the PDB can help with this. 
	- Also, many PDB entries are [missing portions of the molecule](https://pdb101.rcsb.org/learn/guide-to-understanding-pdb-data/missing-coordinates) that were not observed in the experiment. These include structures that include only alpha carbon positions, structures with missing loops, structures of individual domains, or subunits from a larger molecule. 
	- In addition, most of the crystallographic structure entries do not have information on hydrogen atoms.
- Static conformation in crystallizing state
	- Many different conformations
	- Probably different structure in a cell