# Advanced structure prediction

[0. VU Structural Bioinformatics](Biology/VU%20Structural%20Bioinformatics/0.%20VU%20Structural%20Bioinformatics.md)

## Protein folding and protein structure

- Sequence to 3d structure of protein
- Challenges and open questions:
	- Mechanisms: how does protein fold automatically into its structure; free energy landscape
	- Driving forces: what guides the molecule into its biologically relevant state
	- Protein dynamics & function: proteins can take different conformations

## Experimental structure determination

[2. VU SB Structure Determination](Biology/VU%20Structural%20Bioinformatics/2.%20VU%20SB%20Structure%20Determination.md)

| Method                            | Pro                                                                 | Con                                                                                 |
| --------------------------------- | ------------------------------------------------------------------- | ----------------------------------------------------------------------------------- |
| X-ray crystallography             | - High resolution<br>- No size limits<br>- Easy addition of ligands | - Crystals required<br>- Phase problem                                              |
| Electron microscopy / diffraction | - Single particles possible<br>- Well suited to membrane proteins   | - Labor intensive<br>- High resolution difficult                                    |
| NMR spectroscopy                  | - In solution<br>- Dynamic formation possible                       | - Resolution variable<br>- Limited size (< ~30 kDa)                                 |
| Other spectroscopy                | - (Very) high time resolution<br>- (Very) sensitive                 | - Technically extremely complex<br>- Limited applicability<br>- Limited information |
- Expensive
- Time consuming
- Require additional data processing and modeling
- Not accessible for all proteins

## The Nobel Prize in chemistry 2024

- DeepMind: Demis Hassabis, John M. Jumper
- University of Washington: David Baker
- Computational protein design and protein structure prediction

## Structure prediction before AF

[4. VU SB Introduction to Structure Prediction](Biology/VU%20Structural%20Bioinformatics/4.%20VU%20SB%20Introduction%20to%20Structure%20Prediction.md)

- Homology modeling
	- From homologous sequences
	- Build homology model
- No template
	- Build from fragments

### Homology modeling

- Minimal sequence identity for it to work: ~20-25%
	- Below this molecular dynamics modeling using the model may not work
- Good: > 40-50%
- 20-30% for GPCRs (gene protein coupled receptors)

### CASP

[Critical Assessment of Protein Structure Prediction](Biology/VU%20Structural%20Bioinformatics/4.%20VU%20SB%20Introduction%20to%20Structure%20Prediction.md#Critical%20Assessment%20of%20Protein%20Structure%20Prediction)

- Why is it useful and how it leads to the success of AF
	- Unbiased comparison of performance
	- Compare different approaches to predict structure
- What makes a good benchmark
	- Clearly defined task
	- Independent assessment
	- Golden standard: experimentally determined structures
	- Metric: [GDT_TS](Biology/VU%20Structural%20Bioinformatics/4.%20VU%20SB%20Introduction%20to%20Structure%20Prediction.md#GDT_TS)
		- RMSD is sensitive to outliers
		- GDT_TS > 90% is considered equivalent to experimentally determined structure
- CASP 2020: AF2 reaching nearly 90% GDT_TS

## Neural networks for structure predictions

- AF2: From sequence to protein structure
- Based on neural nets

### Inductive biases included in DL models

- The structure of the NN is designed to suit the problem to solve
- Image processing: CNN
- Small molecule: GNN
- Language: RNN
- AF2: Attention

### AF1

- From CNN
- Made distance predictions and angle predictions
- Calculates a score -> Gradient descent
- Outputs structure

### AF2

- More complex NN architecture
- Uses:
	- Physical and biological knowledge about protein structure
		- evolution of protein
		- geometry of residues
		- not really physics like thermodynamics of folding
	- multiple sequence alignment 
	- deep learning
- Information flow:
	- Database search & pre-processing
	- Evoformer
	- Structure module
	- Recycling: feed output structure back to model to improve prediction

## AF2

![](Media/Pasted%20image%2020250219121647.png)

### Database search & pre-processing

- **Inputs**: 
	- sequence
	- genetic database: for MSA
	- pairing: for pair representation
	- structure database: to find a structure template (optional)
- **Outputs**:
	- Evolutionary information through **MSA**
		- Conserved amino acid
		- Mutation
		- Co-evolution: when 1 AA mutates, the AA it bonds with in the structure also needs to mutate
	- Pair representation (representation between pairs of AAs)
		- Characterize the structure
		- Can encode distances or angles

### Recycling

- Makes an initial predicted structure
- Feed the predicted structure back to the network
- The network improves the predicted structure in iterations until convergence

### Evoformer

- **Inputs**:
	- MSA representation (s,r,c)
	- Pair representation (r,r,c)
- **Outputs**: refined representations
	- Single representation (r,c)
	- Pair representation (r,r,c)
- Architecture based on attention mechanism
- Communication between MSA and pair representation
- Pair information: used for **triangle constraint**
	- If distances between AB and AC are known, it places a constraint on BC.

### Structure module

- **Inputs**:
	- Single representation (r,c)
	- Pair representation (r,r,c)
- **Output**: 
	- atomic coordinates of the 3D structure
	- Metric to identify regions of high/low confidence
- Performs _rotations and translations_ on each AA
- Considers physical and chemical constraints
- **Residue gas**:
	- traditionally the peptide chain imposes a constraint on the coordinates of the AAs
	- The residue gas model removes the constraint from the peptide chain, treating each residue as a free-floating rigid body
	- enables _local optimization_
	- avoids loop closure problem by not considering the peptide chain
	- iteratively refine the predictions
- **Frame aligned point error (FAPE)**: an error function

### AMBER force field

- Assisted Model Building with Energy Refinement (AMBER)
- A set of empirical rules and mathematical formulae used in molecular dynamics simulations to predict the behavior of molecules, particularly proteins and nucleic acids
- After AF2 predicts the 3D coordinates of a protein, the AMBER force field helps to optimize these coordinates by _minimizing the energy of the system_, ensuring that the predicted structure is _physically plausible and stable_

### Noisy student distillation

- The model is first trained on labeled (with structure) data from [PDB](Biology/VU%20Structural%20Bioinformatics/1.%20VU%20SB%20Introduction.md#PDB) (_teacher model_)
- The teacher model _predicts_ structures for a set of unlabeled sequences
- A new dataset of _high confidence predicted structures_ is created from the predictions
- A student model is trained from the teacher model, using a combination of _labeled and predicted data_. Noise is added to make the model learn more robust features.
- The process is repeated multiple times (student model becomes teach) to improve accuracy.

## Access to AF2

- Open source, contains predicted structures of the entire human proteome
- pLDDT score:
	- > 90: good
	- > 70: lower confidence
	- < 50: a disorder prediction
- Keep in mind that AF2 can give wrong structures
- Impacts of AF2:
	- Breakthrough in protein structure prediction
	- Accelerated scientific research
	- Open-access database: over 200 million protein structures available to researchers
	- Demonstrated the potential of AI in solving complex scientific problems
	- **Applications**:
		- Drug design
		- Enzyme engineering
		- Structural biology
		- Link omics to structure

## Explainability: understanding the underlying mechanisms

- How does the model make decisions
	- DL is pretty much a black box
- What can be learned about the underlying biology
	- Understanding protein mis-folding & [Aggregation](Biology/VU%20Structural%20Bioinformatics/4.%20VU%20SB%20Introduction%20to%20Structure%20Prediction.md#Aggregated%20state)
- What are the limitations of the AI models?
	- There may be unknown secondary structures: not helices, sheets, or other known structures

### Deep MSA

- The authors demonstrated having a deeper MSA (more sequences aligned in MSA) improves the accuracy.
- The accuracy scores stabilize at around $N_{eff}\approx 100$ 

## Next steps

- Using deep learning
	- De novo molecule design
	- Mutation impact prediction
	- More
- Explainability and deep learning models
	- Understanding model choices during predictions
	- Leveraging DL to understand biological data
	- Make sense of omics data to understand biological conditions
		- Use AI to understand the data, not just learn from it
- FAIR data and workflows
	- Crucial for innovation
	- Data sensitivity
- AF3: already out