# Structure Comparison

[0. VU Structural Bioinformatics](Biology/VU%20Structural%20Bioinformatics/0.%20VU%20Structural%20Bioinformatics.md)
chapter 3 in book

## Structure vs. sequence comparison

- Similarity between 2 proteins can be compared by a [pair-wise alignment](Biology/MIT%20Computational%20Biology/3.%20MIT%20CompBio%20-%20Hashing,%20Database%20Search,%20BLAST%20algorithm.md) of 2 sequences
- However, this works well only for 2 proteins that are close homologs
- A structural comparison through _structural alignment_ gives much more valuable evolutionary information
	- distant evolutionary relationships are more easily found via structural comparisons
- Two similar structures can have very low sequence similarity
- Challenge in structural alignment: there are many ways two structures can be aligned

## Structure comparison

- Structures are more evolutionarily conserved than sequences are
	- changes in structures have a more profound impact so are more selected against
- **Question**: how similar are two structures
- Structure _superposition_
	- a solved problem
	- need an _alignment_ between each pair of residues in the two structures
	- ex. superimpose the predicted and the solution structures of the same protein. Since they have the same amino acid sequences, the alignment is there already
- Structure alignment
	- analogous to [sequence alignments](Biology/MIT%20Computational%20Biology/2.%20MIT%20CompBio%20-%20Dynamic%20Programming.md), but different
	- two very different sequences can have similar structures
	- align different residues in two different structures

### PDB file

- Stores protein structure
- Each atom's x, y, and z coordinates are recorded
- Note that the protein can be _rotated_ and _translated_ without changing the interatomic distances

### Structure superposition

- Input:
	- 2 proteins with atomic coordinates (2 pdb files)
	- a mapping indicating which are the corresponding residues
- Output:
	- 2 proteins superimposed typically by providing a _rotation_ and _translation_ of the coordinate frame
- $O(N^P)$: polynomial

### Structure alignment

- Input:
	- 2 proteins with their atomic coordinates (2 PDB files)
- Output:
	- an alignment between 2 protein structures based on the _structure alone_
	- the program often also outputs a superposition using the alignment
	- a structural alignment score
- NP hard: non-polynomial; heuristic needed; see [NP Problems](Machine%20Learning/Concepts/NP%20Problems.md)

## Measure of difference

### RMSD

- The RMSD is defined as: $$\begin{aligned}\mathrm{RMSD}(V,W)&=\sqrt{\frac{1}{n}\sum_{i\in M}\|\mathbf{v_i}-\mathbf{w_i}\|^2}\\&=\sqrt{\frac{1}{n}\sum_{i\in M}(v_{i,x}-w_{i,x})^2+(v_{i,y}-w_{i,y})^2+(v_{i,z}-w_{i,z})^2}\end{aligned}$$, where $n$ is the number of aligned residues and $M$ is the number of matched pairs of atoms between structure $V$ and $W$

### RMSD for superposition

- _given_ how residues map, find rotation and translation that minimizes the RMSD.
- find translation so that **centers of mass** fall on each other:
	- Center of mass: $\mathbf{R}=\frac{\sum^{i=N}_{i=0}m_ir_i}{\sum^{i=N}_{i=0}m_i}$, where $m_i$ is the mass of each atom $i$, $r_i$ is the atomic coordinates of each atom $i$, and $N$ is the total number of atoms
- rotation is found with the Jacobi algorithm to solve eigenvalue problem in polynomial time (Kearsly 1989).
- Then calculate the RMSD between them.
 
### RMSD for alignment

- 2 different structures
- if we are to calculate the RMSD between the 2 structures, their structural alignment needs to be found first
- this cannot be solved with aligning _residues_ using dynamic programming because structures can be more flexible
	- sequences cannot be swapped in order, but structures can
	- the optimal global solution does not consist of optimal solutions to the sub-problems
- one way to do it: aligning using [phi-psi angles](Biology/VU%20Structural%20Bioinformatics/3.%20VU%20SB%20Protein%20Folding.md#Dihedral%20angles). If two structures share similar sequences of phi-psi angles, they are similar.
	- has a problem: for example, two different helices can have similar sequence of angles

## Considering sequence alignment

- NP problem: need to find a heuristic solution
- **Representation**
	- sequence and their atomic coordinates
- **Optimization**
	- dynamic programming: [Needleman-Wnusch algorithm](Biology/MIT%20Computational%20Biology/2.%20MIT%20CompBio%20-%20Dynamic%20Programming.md#Dynamic%20Programming%20for%20Sequence%20Alignment) / [Smith-Waterman Algorithm](Biology/MIT%20Computational%20Biology/3.%20MIT%20CompBio%20-%20Hashing,%20Database%20Search,%20BLAST%20algorithm.md#Smith-Waterman%20Algorithm)
- **Scoring**
	- RMSD: slow because each calculation uses and changes the whole alignment
	- need to define a local similarity measure
	- need to be coordinate independent such that it is invariant to rotations and translations
	- use distance: _contact matrix_

### Contact matrices

![](Media/Pasted%20image%2020250224102123.png)
- Shows whether there is contact between pairs of residues
- Diagonal: same residue
- Other space: distance between the structures
- Rotation/translation invariant
- Indicates local structural neighborhood of each residue

### Recognizing helix and sheets

![[Media/7. VU SB Structure Comparison 2025-02-24 10.26.46.excalidraw]]

- On the contact matrix, the closer the points are to the diagonal, the closer the contact distance
- **Alpha helix**: the interactions are between $i$ and $i+3$. So the points are close to the diagonal. Also, the contact distance is between pairs of residues is always 3.
- **Parallel strands**: the interactions are between $i$ and $i+k$, where $k$ is a constant. In the illustration above, $k$ is 14. Since the interaction is distance is longer, the points are further from the diagonal. The contact distance is alway $k$.
- **Anti-parallel strands**: the interactions are between $i+n$ and $j-n$, where n is an integer between 1 and the length of a strand - 1. So as $i$ increases, $j$ decreases. The points form a line with negative slope. Anti-parallel strands are also more distant interactions, so the points are further from the diagonal.

### Contact matrix formula

- - Given this definition for a contact matrix
$$C_{i,j}=
\begin{Bmatrix}
1 & \text{if i and j make a contact} \\
0 & \mathrm{otherwise}
\end{Bmatrix}$$
- Define a score: $$CO(V)=\frac{\sum_{i=0}^{i=N}\sum_{j=i}^{j=N}|j-i|C_{ij}}{\sum_{i=0}^{i=N}\sum_{j=i}^{j=N}C_{ij}}$$
- This measure
	- Bottom: total number of contacts
	- Top: total contact distance = $\text{number of contacts}\times \text{interaction distances}$ 
	- Together: average contact distance

### Double dynamic programming

- Need some estimate of alignment first to do alignment